// @alifd/meet/lib/date-picker/embed-picker/date-picker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var dayjs_1 = tslib_1.__importDefault(require("dayjs"));

var month_picker_1 = tslib_1.__importDefault(require("./month-picker"));

var utils_1 = require("../../utils");

var utils_2 = require("./utils");

var EmbedDatePicker = function (_super) {
  tslib_1.__extends(EmbedDatePicker, _super);

  function EmbedDatePicker() {
    return _super !== null && _super.apply(this, arguments) || this;
  }

  EmbedDatePicker.prototype.getStep = function () {
    return 'day';
  };

  EmbedDatePicker.prototype.getDisableType = function () {
    return 'day';
  };

  EmbedDatePicker.prototype.getRangeCheck = function () {
    return {
      date: {
        hasYear: true,
        hasMonth: true,
        hasDay: true
      }
    };
  };

  EmbedDatePicker.prototype.getDataSource = function (locale) {
    var result = _super.prototype.getDataSource.call(this, locale);

    var dayLocale = utils_1.get(locale, 'locale.DatePicker.day', '日');
    var month = this.value.getMonth();
    var year = this.value.getFullYear();
    var days = dayjs_1.default(year + "-" + (month + 1) + "-01").daysInMonth();
    result.push(this.filterDataSource(utils_2.rangeOptions(1, days, dayLocale), function (d, it) {
      utils_2.resetTime(d);
      d.setFullYear(year);
      d.setMonth(month);
      d.setDate(it.value);
    }, 'day'));
    return result;
  };

  EmbedDatePicker.prototype.toDate = function (values) {
    var d = _super.prototype.toDate.call(this, values);

    d.setDate(values[2]);
    return d;
  };

  EmbedDatePicker.prototype.toValues = function (d) {
    var values = _super.prototype.toValues.call(this, d);

    values.push(d.getDate());
    return values;
  };

  return EmbedDatePicker;
}(month_picker_1.default);

exports.default = EmbedDatePicker;
