// @alilc/lowcode-material-parser/lib/generate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.genManifest = void 0;
const tslib_1 = require("tslib");
const path = tslib_1.__importStar(require("path"));
const core_1 = require("./core");
const log = core_1.debug.extend('gen');
async function default_1(matScanModel, matParsedModels, options) {
    const containerList = [];
    for (const matParsedModel of matParsedModels) {
        // 默认排除掉 defaultExportName 为空的组件
        if (!matParsedModel.componentName) {
            log('skip');
            continue;
        }
        // 组装 manifest
        const manifest = await genManifest(matScanModel, matParsedModel, options);
        containerList.push(manifest);
    }
    return containerList;
}
exports.default = default_1;
/**
 * 生成 manifest
 *
 * @param {IMaterialParsedModel} matParsedModel
 * @returns {Promise<
 *     manifestObj: ComponentMeta, // 组件描述
 *   >}
 * @memberof LocalGenerator
 */
async function genManifest(matScanModel, matParsedModel, options) {
    var _a, _b, _c;
    const manifestObj = {
        componentName: matParsedModel.componentName,
        title: matScanModel.pkgName,
        docUrl: '',
        screenshot: '',
        devMode: 'proCode',
        npm: {
            package: matScanModel.pkgName,
            version: matScanModel.pkgVersion,
            exportName: ((_a = matParsedModel.meta) === null || _a === void 0 ? void 0 : _a.exportName) || matParsedModel.componentName,
            main: options.root && path.isAbsolute(matScanModel.mainFilePath)
                ? path.relative(options.root, matScanModel.mainFilePath)
                : matScanModel.mainFilePath,
            destructuring: ((_b = matParsedModel.meta) === null || _b === void 0 ? void 0 : _b.exportName) !== 'default',
            subName: ((_c = matParsedModel.meta) === null || _c === void 0 ? void 0 : _c.subName) || '',
        },
    };
    // 填充 props
    manifestObj.props = matParsedModel.props;
    // 执行扩展点
    return manifestObj;
}
exports.genManifest = genManifest;
