// @alilc/lowcode-rax-renderer/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.Engine = void 0;
var _rax = require("rax");
var _raxFindDomNode = _interopRequireDefault(require("rax-find-dom-node"));
var _lowcodeRendererCore = require("@alilc/lowcode-renderer-core");
var _page = _interopRequireDefault(require("./renderer/page"));
var _component = _interopRequireDefault(require("./renderer/component"));
var _block = _interopRequireDefault(require("./renderer/block"));
var _compFactory = _interopRequireDefault(require("./hoc/compFactory"));
exports.CompFactory = _compFactory.default;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
_lowcodeRendererCore.adapter.setRuntime({
  Component: _rax.Component,
  PureComponent: _rax.PureComponent,
  createContext: _rax.createContext,
  createElement: _rax.createElement,
  forwardRef: _rax.forwardRef,
  findDOMNode: _raxFindDomNode.default
});
_lowcodeRendererCore.adapter.setRenderers({
  PageRenderer: (0, _page.default)(),
  ComponentRenderer: (0, _component.default)(),
  BlockRenderer: (0, _block.default)(),
  AddonRenderer: (0, _lowcodeRendererCore.addonRendererFactory)(),
  TempRenderer: (0, _lowcodeRendererCore.tempRendererFactory)()
});
function factory() {
  var Renderer = (0, _lowcodeRendererCore.rendererFactory)();
  return /*#__PURE__*/function (_Renderer) {
    _inheritsLoose(_class, _Renderer);
    function _class(props, context) {
      return _Renderer.call(this, props, context) || this;
    }
    var _proto = _class.prototype;
    _proto.isValidComponent = function isValidComponent(obj) {
      var _obj$prototype;
      return (obj == null ? void 0 : (_obj$prototype = obj.prototype) == null ? void 0 : _obj$prototype.setState) || (obj == null ? void 0 : obj.prototype) instanceof _rax.Component;
    };
    return _class;
  }(Renderer);
}
var RaxRenderer = factory();
var Engine = RaxRenderer;
exports.Engine = Engine;
var _default = RaxRenderer;
exports.default = _default;
