// @alifd/meet/lib/transfer/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setCheckedItem = exports.changeCheckedItem = exports.setListDisabled = exports.checkAllData = exports.getDataStatus = exports.filterTransferValue = void 0;

var tslib_1 = require("tslib");

exports.filterTransferValue = function (dataSource, defaultValue) {
  var leftList = [];
  var rightList = [];
  dataSource.map(function (item) {
    if (defaultValue.indexOf(item.value) === -1) {
      leftList.push(item);
    } else {
      rightList.push(item);
    }
  });
  return {
    leftList: leftList,
    rightList: rightList
  };
};

exports.getDataStatus = function (dataSource) {
  var listDataSize = dataSource.length;
  var checkedSize = 0;
  var disabledSize = 0;
  var isCheckAll = false;
  dataSource.map(function (item) {
    if (item.isChecked) {
      checkedSize = checkedSize + 1;
    }

    if (item.disabled) {
      disabledSize = disabledSize + 1;
    }
  });
  isCheckAll = listDataSize - disabledSize === checkedSize && checkedSize > 0 ? true : false;
  return {
    listDataSize: listDataSize,
    checkedSize: checkedSize,
    disabledSize: disabledSize,
    isCheckAll: isCheckAll
  };
};

exports.checkAllData = function (dataSource, status) {
  dataSource.map(function (item) {
    if (!item.disabled) {
      item.isChecked = status;
    }
  });
  return dataSource;
};

exports.setListDisabled = function (dataSource) {
  dataSource.map(function (item) {
    item.disabled = true;
  });
  return dataSource;
};

exports.changeCheckedItem = function (initList, goalList) {
  var initValue = [];
  var changeListValue = [];
  var changeList = [];
  initList = initList.filter(function (item) {
    var isChecked = item.isChecked,
        disabled = item.disabled,
        value = item.value;

    if (isChecked && !disabled) {
      item.isChecked = false;
      changeList.push(item);
      changeListValue.push(value);
    } else {
      initValue.push(value);
      return item;
    }
  });
  goalList = tslib_1.__spreadArrays(changeList, goalList);
  var goalValue = goalList.map(function (item) {
    return item.value;
  });
  return {
    initList: initList,
    initValue: initValue,
    goalList: goalList,
    goalValue: goalValue,
    changeList: changeList,
    changeListValue: changeListValue
  };
};

exports.setCheckedItem = function (dataSource, checkedList) {
  dataSource.map(function (item) {
    if (checkedList.indexOf(item.value) !== -1) {
      item.isChecked = true;
    }
  });
  return dataSource;
};
