// @alifd/meet/lib/range/dom.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getComponentRect = exports.px2rem = void 0;

var tslib_1 = require("tslib");

var utils_1 = require("../utils");

function px2rem(px) {
  return px * 750 / utils_1.windowWidth;
}

exports.px2rem = px2rem;

function getComponentRect(id, callback) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    var rect;
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          if (!callback) return [3, 2];
          return [4, utils_1.getElementBoundingClientRect(id)];

        case 1:
          rect = _a.sent();
          callback({
            result: true,
            size: {
              x: px2rem(rect.x),
              y: px2rem(rect.y),
              width: px2rem(rect.width),
              height: px2rem(rect.height),
              top: px2rem(rect.top),
              bottom: px2rem(rect.bottom),
              left: px2rem(rect.left),
              right: px2rem(rect.right)
            }
          });
          _a.label = 2;

        case 2:
          return [2];
      }
    });
  });
}

exports.getComponentRect = getComponentRect;
