// whale-component-docgen/lib/utils/parseJSComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseJSComponent = void 0;
const docgen = require("react-docgen");
const fs = require("fs-extra");
const whale_docgen_ts_1 = require("whale-docgen-ts");
const parseJSComponent = (componentPath) => {
    const re = docgen.parse(fs.readFileSync(componentPath));
    const docs = Array.isArray(re) ? re : [re];
    return docs.map((vo) => documentationObject2ComponentDoc(vo));
};
exports.parseJSComponent = parseJSComponent;
const generatePropType = (prop) => {
    const type = prop.tsType || prop.flowType || prop.type;
    if (type && 'value' in type && Array.isArray(type.value)) {
        const typeDoc = {
            name: 'union',
            raw: 'raw' in type
                ? type.raw
                : type.value
                    .map((typeValue) => typeValue.name || typeValue.value)
                    .join(' | '),
            value: type.value.map((vo) => whale_docgen_ts_1.createWmPlainPropItemType(vo.value)),
        };
        return typeDoc;
    }
    return whale_docgen_ts_1.createWmPlainPropItemType((type === null || type === void 0 ? void 0 : type.name) || 'unknown');
};
const documentationObject2ComponentDoc = ({ props, displayName, description, }) => ({
    meta: {
        exportName: displayName,
    },
    displayName,
    description,
    props: Object.keys(props).reduce((p, name) => {
        const prop = props[name];
        p[name] = {
            name,
            required: !!prop.required,
            type: generatePropType(prop),
            description: prop.description || '',
            defaultValue: prop.defaultValue || '',
        };
        return p;
    }, {}),
    methods: [],
});
