// @alifd/meet/lib/utils/element.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.scrollToId = exports.getRight = exports.getBottom = exports.getTop = exports.getLeft = exports.getHeight = exports.getWidth = exports.getElementBoundingClientRect = exports.getScrollTop = exports.getScrollLeft = exports.getElementScrollOffset = void 0;

var tslib_1 = require("tslib");

var universal_element_1 = require("universal-element");

var page_scroll_to_1 = tslib_1.__importDefault(require("@uni/page-scroll-to"));

var utils_1 = require("../utils");

function getElementScrollOffset(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    return tslib_1.__generator(this, function () {
      return [2, universal_element_1.getScrollOffset("#" + id).then(function (ret) {
        var offset = {
          scrollLeft: 0,
          scrollTop: 0
        };

        if (ret && ret[0]) {
          var _a = ret[0],
              scrollLeft = _a.scrollLeft,
              scrollTop = _a.scrollTop;
          return tslib_1.__assign(tslib_1.__assign({}, offset), {
            scrollLeft: scrollLeft,
            scrollTop: scrollTop
          });
        }

        return offset;
      })];
    });
  });
}

exports.getElementScrollOffset = getElementScrollOffset;

function getScrollLeft(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          return [4, getElementScrollOffset(id)];

        case 1:
          return [2, _a.sent().scrollLeft];
      }
    });
  });
}

exports.getScrollLeft = getScrollLeft;

function getScrollTop(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          return [4, getElementScrollOffset(id)];

        case 1:
          return [2, _a.sent().scrollTop];
      }
    });
  });
}

exports.getScrollTop = getScrollTop;

function getElementBoundingClientRect(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    var getRect, rs;
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          getRect = universal_element_1.getBoundingClientRect;

          if (utils_1.isStrictWechatMiniProgram) {
            getRect = function (selector) {
              return new Promise(function (resolve) {
                var ele = document.querySelector(selector);

                if (!ele || !ele._internal) {
                  if (process.env.NODE_ENV === 'development') {
                    console.warn('element not found', selector);
                  }

                  resolve([]);
                }

                ele._internal.createSelectorQuery().select(selector).boundingClientRect().exec(function (res) {
                  resolve(res);
                });
              });
            };
          }

          return [4, getRect("#" + id).then(function (ret) {
            var rect = {
              left: 0,
              bottom: 0,
              top: 0,
              right: 0,
              width: 0,
              height: 0,
              x: 0,
              y: 0
            };

            if (ret && ret[0]) {
              var _a = ret[0],
                  left = _a.left,
                  bottom = _a.bottom,
                  top_1 = _a.top,
                  right = _a.right,
                  width = _a.width,
                  height = _a.height,
                  x = _a.x,
                  y = _a.y;
              return tslib_1.__assign(tslib_1.__assign({}, rect), {
                left: left,
                bottom: bottom,
                top: top_1,
                right: right,
                width: width,
                height: height,
                x: x,
                y: y
              });
            }

            return rect;
          })];

        case 1:
          rs = _a.sent();
          return [2, rs];
      }
    });
  });
}

exports.getElementBoundingClientRect = getElementBoundingClientRect;

function getWidth(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          return [4, getElementBoundingClientRect(id)];

        case 1:
          return [2, _a.sent().width];
      }
    });
  });
}

exports.getWidth = getWidth;

function getHeight(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          return [4, getElementBoundingClientRect(id)];

        case 1:
          return [2, _a.sent().height];
      }
    });
  });
}

exports.getHeight = getHeight;

function getLeft(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          return [4, getElementBoundingClientRect(id)];

        case 1:
          return [2, _a.sent().left];
      }
    });
  });
}

exports.getLeft = getLeft;

function getTop(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          return [4, getElementBoundingClientRect(id)];

        case 1:
          return [2, _a.sent().top];
      }
    });
  });
}

exports.getTop = getTop;

function getBottom(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          return [4, getElementBoundingClientRect(id)];

        case 1:
          return [2, _a.sent().bottom];
      }
    });
  });
}

exports.getBottom = getBottom;

function getRight(id) {
  return tslib_1.__awaiter(this, void 0, void 0, function () {
    return tslib_1.__generator(this, function (_a) {
      switch (_a.label) {
        case 0:
          return [4, getElementBoundingClientRect(id)];

        case 1:
          return [2, _a.sent().right];
      }
    });
  });
}

exports.getRight = getRight;

function scrollToId(id) {
  page_scroll_to_1.default({
    selector: "#" + id
  }).then(utils_1.noop);
}

exports.scrollToId = scrollToId;
