// @alifd/meet/lib/rich-text/image.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var rax_image_1 = tslib_1.__importDefault(require("rax-image"));

var classnames_1 = tslib_1.__importDefault(require("classnames"));

var utils_1 = require("../utils");

var RichTextImage = function (props) {
  var item = props.item,
      onClick = props.onClick,
      className = props.className,
      lazyLoad = props.lazyLoad,
      errorImage = props.errorImage,
      imgIndex = props.imgIndex,
      src = props.src,
      onError = props.onError,
      style = props.style,
      otherProps = tslib_1.__rest(props, ["item", "onClick", "className", "lazyLoad", "errorImage", "imgIndex", "src", "onError", "style"]);

  var _a = rax_1.useState(src),
      imgSrc = _a[0],
      setImgSrc = _a[1];

  return rax_1.createElement(rax_image_1.default, tslib_1.__assign({}, otherProps, {
    className: classnames_1.default(className, {
      '_img-error': imgSrc === errorImage
    }),
    source: {
      uri: imgSrc
    },
    mode: "widthFix",
    placeholder: 'https://gw.alicdn.com/imgextra/i4/O1CN01iIbyHS1dSDboBiORe_!!6000000003734-2-tps-200-200.png',
    loading: lazyLoad ? 'lazy' : 'eager',
    style: style,
    onClick: function handleClick() {
      var args = [];

      for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
      }

      if (!imgSrc.includes('data:')) {
        onClick.apply(void 0, args);
      }
    },
    onError: function handleImageErorr() {
      setImgSrc(errorImage);

      if (utils_1.isFunction(onError)) {
        onError(src, imgIndex);
      }
    }
  }));
};

exports.default = RichTextImage;
