// @alifd/meet/lib/rich-text/link.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var navigate_1 = tslib_1.__importDefault(require("@uni/navigate"));

var clipboard_1 = require("@uni/clipboard");

var message_1 = tslib_1.__importDefault(require("../message"));

var utils_1 = require("../utils");

var MINIAPP_PAGE_REG = /^\/[a-zA-z0-9]/;
var SIMPLE_URL_REG = /[a-zA-z]+:\/\/[^\s]*/;

var RichTextLink = function (props) {
  var style = props.style,
      href = props.href,
      className = props.className,
      children = props.children,
      onLinkClick = props.onLinkClick,
      otherProps = tslib_1.__rest(props, ["style", "href", "className", "children", "onLinkClick"]);

  return rax_1.createElement("view", tslib_1.__assign({}, otherProps, {
    className: className,
    onClick: function handleClick() {
      if (href[0] === '#') {
        var id = href.substring(1);
        utils_1.scrollToId(id);
      } else if (SIMPLE_URL_REG.test(href)) {
        if (utils_1.isFunction(onLinkClick) && onLinkClick(href) !== false || utils_1.isUndef(onLinkClick)) {
          clipboard_1.setClipboard({
            text: href
          }).then(function () {
            if (utils_1.isStrictMiniApp || utils_1.isStrictByteDanceMicroApp) {
              message_1.default.show({
                content: '链接已复制到粘贴板'
              });
            }
          });
        }
      } else if (MINIAPP_PAGE_REG.test(href)) {
        navigate_1.default.push({
          url: href
        }).then(function () {});
      }
    },
    style: style
  }), children);
};

exports.default = RichTextLink;
