// rax/lib/vdom/component.js
"use strict";

exports.__esModule = true;
exports.PureComponent = exports.default = void 0;

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

/**
 * Base component class.
 */
var Component = /*#__PURE__*/function () {
  function Component(props, context) {
    this.props = props;
    this.context = context;
    this.refs = {};
  }

  var _proto = Component.prototype;

  _proto.setState = function setState(partialState, callback) {
    // The updater property is injected when composite component mounting
    this.updater.setState(this, partialState, callback);
  };

  _proto.forceUpdate = function forceUpdate(callback) {
    this.updater.forceUpdate(this, callback);
  };

  return Component;
}();
/**
 * Pure component.
 */


exports.default = Component;

var PureComponent = /*#__PURE__*/function (_Component) {
  _inheritsLoose(PureComponent, _Component);

  function PureComponent(props, context) {
    var _this;

    _this = _Component.call(this, props, context) || this;
    _this.__isPureComponent = true;
    return _this;
  }

  return PureComponent;
}(Component);

exports.PureComponent = PureComponent;
