// rax/es/index.js
import _createElement from './createElement';
export { _createElement as createElement };
import _createContext from './createContext';
export { _createContext as createContext };
import _createRef from './createRef';
export { _createRef as createRef };
import _forwardRef from './forwardRef';
export { _forwardRef as forwardRef };
export { useState, useContext, useEffect, useLayoutEffect, useRef, useCallback, useMemo, useReducer, useImperativeHandle } from './hooks';
import _memo from './memo';
export { _memo as memo };
import _Fragment from './fragment';
export { _Fragment as Fragment };
import _render from './render';
export { _render as render };
import _Component from './vdom/component';
export { _Component as Component };
export { PureComponent } from './vdom/component';
import _version from './version';
export { _version as version };
import Host from './vdom/host';
import Instance from './vdom/instance';
import Element from './vdom/element';
import flattenChildren from './vdom/flattenChildren';
import DevtoolsHook from './devtools/index';
export var shared = {
  Host: Host,
  Instance: Instance,
  Element: Element,
  flattenChildren: flattenChildren
};

if (process.env.NODE_ENV !== 'production') {
  /* global __RAX_DEVTOOLS_GLOBAL_HOOK__ */
  if (typeof __RAX_DEVTOOLS_GLOBAL_HOOK__ !== 'undefined' && typeof __RAX_DEVTOOLS_GLOBAL_HOOK__.inject === 'function') {
    __RAX_DEVTOOLS_GLOBAL_HOOK__.inject(DevtoolsHook);
  }

  if (typeof window !== 'undefined') {
    if (window.__RAX_INITIALISED__) {
      console.error('Warning: more than one instance of Rax has been initialised, this could lead to unexpected behaviour.');
    }

    window.__RAX_INITIALISED__ = true;
  }
}
