// @alifd/meet/lib/upload/request.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var file_1 = tslib_1.__importDefault(require("@uni/file"));

var utils_1 = require("../utils");

var utils_2 = require("./utils");

function request(obj) {
  var task = file_1.default.upload({
    url: obj.action,
    formData: obj.data,
    filePath: obj.file,
    withCredentials: obj.withCredentials,
    fileName: obj.filename,
    fileType: obj.fileType,
    header: obj.headers,
    success: function success(res) {
      return obj.onSuccess(res);
    },
    fail: function fail(e) {
      return obj.onError(e);
    }
  });
  var skipTaskProgressUpdate = utils_1.isAndroid && utils_1.isStrictWechatMiniProgram;

  if (task && task.onProgressUpdate && !skipTaskProgressUpdate) {
    task.onProgressUpdate(function (_a) {
      var progress = _a.progress;
      obj.onProgress({
        percent: progress
      });
    });
  } else {
    var percent_1 = 10;
    var timer_1 = setInterval(function () {
      if (percent_1 < 90) {
        percent_1 += 10;
        obj.onProgress({
          percent: percent_1
        });
      } else {
        clearInterval(timer_1);
      }
    }, utils_2.UPLOAD_THROTTLE);
  }

  return {
    abort: function abort() {
      if (task && task.abort) {
        task.abort();
      }
    }
  };
}

exports.default = request;
