// @alifd/meet/lib/upload/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.chooseImageForWechat = exports.createFileItem = exports.UPLOAD_THROTTLE = void 0;

var utils_1 = require("../utils");

exports.UPLOAD_THROTTLE = 800;

function createFileItem(url, file) {
  var id = Math.ceil(Math.random() * 1e7);
  var isImage = utils_1.isMiniappPlatform || url.startsWith('data:image/');
  return {
    uid: "file_" + id,
    name: 'file',
    state: 'selected',
    response: {},
    url: '',
    percent: 0,
    isImage: isImage,
    file: utils_1.isStrictMiniApp ? url : file || url,
    dataURL: url
  };
}

exports.createFileItem = createFileItem;

function chooseImageForWechat(options) {
  return new Promise(function (resolve, reject) {
    var _a = options.count,
        count = _a === void 0 ? 1 : _a,
        _b = options.mediaType,
        mediaType = _b === void 0 ? ['video', 'image'] : _b,
        _c = options.sourceType,
        sourceType = _c === void 0 ? ['camera', 'album'] : _c,
        _d = options.maxDuration,
        maxDuration = _d === void 0 ? 30 : _d,
        sizeType = options.sizeType;
    wx.chooseMedia({
      sourceType: sourceType,
      count: count,
      maxDuration: maxDuration,
      mediaType: mediaType,
      sizeType: sizeType,
      success: function success(res) {
        var data = [];
        var previewImgs = [];

        if (res.tempFilePaths) {
          data = res.dataPaths;
        } else if (Array.isArray(res.tempFiles)) {
          data = res.tempFiles.map(function (item) {
            return item.tempFilePath || '';
          });
          previewImgs = res.tempFiles.map(function (item) {
            return item.thumbTempFilePath || '';
          });
        }

        resolve(Object.assign({}, {
          data: data,
          previewImgs: previewImgs,
          tempFilePaths: data
        }, res));
      },
      fail: function fail(e) {
        reject(e);
      }
    });
  });
}

exports.chooseImageForWechat = chooseImageForWechat;
