// build-plugin-component-multiple/src/utils/generators/generateMaterialJson.js
const path = require('path');
const fse = require('fs-extra');

const DEV_MOD_MAP = {
  LowCode: 1,
  ProCode: 2,
};

const JS_Library_MAP = {
  react: 1,
  vue: 2,
  rax: 3,
};

const COMPONENT_LIBRARY = {
  next: 1,
  antd: 2,
  meet: 3,
};

const PLATFORM_MAP = {
  pc: 1,
  mobile: 2,
  miniapp: 4,
};

const DSL_MAP = {
  react: 'react',
  rax: 'rax',
  common: 'common',
};

const LOGO_URL = '';

/**
 * 生成物料中心上翻需要的 materials.json
 * 数据结构: https://fusion.alibaba-inc.com/mc/help.html/big-pkg
 */

module.exports = function (context, DSL, components = []) {
  const { pkg, rootDir } = context;
  const { name, version, title, description, homepage } = pkg;

  const jsLibrary = DSL === DSL_MAP.react ? JS_Library_MAP.react : JS_Library_MAP.rax;
  const componentLibrary = DSL === DSL_MAP.react ? COMPONENT_LIBRARY.next : COMPONENT_LIBRARY.meet;
  const platform = DSL === DSL_MAP.react ? PLATFORM_MAP.pc : PLATFORM_MAP.mobile;
  const materialFilePath = path.join(rootDir, 'build/materials.json');

  const json = {
    meta: {
      npm: name,
      name: name,
      version: version,
      title: title,
      description: description,
      homepage: homepage,
      logo: LOGO_URL,
      devMode: DEV_MOD_MAP.ProCode,
      jsLibrary,
      componentLibrary,
      platform,
    },
  };

  json.components = components;

  fse.writeJsonSync(materialFilePath, json, { spaces: 2 });

  return true;
};
