// rax-video/lib/miniapp-runtime/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _rax = require("rax");
var _dedupe = _interopRequireDefault(require("classnames/dedupe"));
var _omit = _interopRequireDefault(require("omit.js"));
var _wrapDefaultProperties = _interopRequireDefault(require("../utils/wrapDefaultProperties"));
var _video = require("@uni/video");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
var miniappVideoPropsMap = {
  showMuteBtn: 'show-mute-btn',
  showPlayBtn: 'show-play-btn',
  showFullscreenBtn: 'show-fullscreen-btn',
  showCenterPlayBtn: 'show-center-play-btn',
  showThinProgressBar: 'show-thin-progress-bar',
  objectFit: 'object-fit'
};
var Video = (0, _rax.forwardRef)(function (props, ref) {
  var id = props.id,
    _props$className = props.className,
    className = _props$className === void 0 ? '' : _props$className,
    style = props.style,
    controls = props.controls,
    playControl = props.playControl,
    autoPlay = props.autoPlay;
  var refEl = (0, _rax.useRef)(null);
  (0, _rax.useImperativeHandle)(ref, function () {
    return refEl.current;
  });
  var common = (0, _omit.default)(props, ['className', 'controls', 'style', 'playControl', 'autoPlay']);
  // Default controls is true
  common.controls = controls === undefined || controls === true;
  common.autoplay = playControl === 'play' || autoPlay;
  Object.keys(miniappVideoPropsMap).forEach(function (prop) {
    common[miniappVideoPropsMap[prop]] = common[prop];
    delete common[prop];
  });
  (0, _rax.useEffect)(function () {
    if (playControl !== undefined) {
      if (!id) {
        console.warn('id is required if using playControl in miniapp!');
        return;
      }
      var cacheVideoContext = function () {
        refEl.current = (0, _video.createVideoContext)(id);
        window.removeEventListener('setDataFinished', cacheVideoContext);
      };
      window.addEventListener('setDataFinished', cacheVideoContext);
    }
  }, []);
  (0, _rax.useEffect)(function () {
    if (playControl !== undefined && refEl.current) {
      playControl === 'play' ? refEl.current.play() : refEl.current.pause();
    }
  }, [playControl]);
  return (0, _rax.createElement)("video", _extends({}, common, {
    className: (0, _dedupe.default)('rax-video', className),
    style: style,
    "webkit-playsinline": true,
    playsinline: true,
    src: props.src
  }));
});
var _default = (0, _wrapDefaultProperties.default)(Video);
exports.default = _default;
