// @alilc/lowcode-types/lib/shell/enum/transform-stage.js
"use strict";

exports.__esModule = true;
exports.TransformStage = exports.IPublicEnumTransformStage = void 0;
var IPublicEnumTransformStage;
/**
 * @deprecated use IPublicEnumTransformStage instead
 */
exports.IPublicEnumTransformStage = IPublicEnumTransformStage;
(function (IPublicEnumTransformStage) {
  IPublicEnumTransformStage["Render"] = "render";
  IPublicEnumTransformStage["Serilize"] = "serilize";
  IPublicEnumTransformStage["Save"] = "save";
  IPublicEnumTransformStage["Clone"] = "clone";
  IPublicEnumTransformStage["Init"] = "init";
  IPublicEnumTransformStage["Upgrade"] = "upgrade";
})(IPublicEnumTransformStage || (exports.IPublicEnumTransformStage = IPublicEnumTransformStage = {}));
var TransformStage;
exports.TransformStage = TransformStage;
(function (TransformStage) {
  TransformStage["Render"] = "render";
  TransformStage["Serilize"] = "serilize";
  TransformStage["Save"] = "save";
  TransformStage["Clone"] = "clone";
  TransformStage["Init"] = "init";
  TransformStage["Upgrade"] = "upgrade";
})(TransformStage || (exports.TransformStage = TransformStage = {}));
