// rax/es/devtools/index.js
import Instance from '../vdom/instance';
import Host from '../vdom/host';
import Reconciler from './reconciler';
import { RENDERED_COMPONENT } from '../constant';
var DevtoolsHook = {
  ComponentTree: {
    getClosestInstanceFromNode: function getClosestInstanceFromNode(node) {
      return Instance.get(node);
    },
    getNodeFromInstance: function getNodeFromInstance(inst) {
      // inst is an internal instance (but could be a composite)
      while (inst[RENDERED_COMPONENT]) {
        inst = inst[RENDERED_COMPONENT];
      }

      if (inst) {
        return inst._nativeNode;
      } else {
        return null;
      }
    }
  },
  Mount: {
    get _instancesByReactRootID() {
      var rootComponents = {}; // Ignore display top-level root component

      for (var rootID in Host.rootComponents) {
        rootComponents[rootID] = Host.rootComponents[rootID][RENDERED_COMPONENT];
      }

      return rootComponents;
    },

    _renderNewRootComponent: Reconciler.renderNewRootComponent
  },
  Reconciler: Reconciler,
  // monitor the info of all components
  monitor: null
};
export default DevtoolsHook;
