// build-plugin-component-multiple/src/configs/rax/getBaseWebpack.js
const webpack = require('webpack');
const path = require('path');
const Chain = require('webpack-chain');
const { setBabelAlias } = require('rax-compile-config');
const getBabelConfig = require('rax-babel-config');
const CaseSensitivePathsPlugin = require('case-sensitive-paths-webpack-plugin');
const TerserPlugin = require('terser-webpack-plugin');
const ProgressPlugin = require('webpackbar');
const OptimizeCSSAssetsPlugin = require('optimize-css-assets-webpack-plugin');
const babelMerge = require('babel-merge');
const { NORMAL_COLOR } = require('../../constants');

module.exports = (context, options = {}) => {
  const { rootDir, command, pkg } = context;
  const { isES6, target, name, progressBarColor = NORMAL_COLOR, enableProgress } = options || {};

  const config = new Chain();

  setBabelAlias(config);

  config.target('web');
  config.context(rootDir);

  config.externals([
    function (ctx, request, callback) {
      if (request.indexOf('@weex-module') !== -1) {
        return callback(null, `commonjs ${request}`);
      }
      // compatible with @system for quickapp
      if (request.indexOf('@system') !== -1) {
        return callback(null, `commonjs ${request}`);
      }
      // compatible with miniapp plugin
      if (/^plugin:\/\//.test(request)) {
        return callback(null, `commonjs ${request}`);
      }
      callback();
    },
  ]);

  config.resolve.extensions.merge(['.js', '.json', '.jsx', '.ts', '.tsx', '.html']);

  const babelConfig = getBabelConfig({
    styleSheet: false,
    isES6,
    custom: {
      ignore: ['**/**/*.d.ts'],
    },
  });
  const mergeBabelConfig = babelMerge(babelConfig, {
    plugins: [require.resolve('cloud-retail-rax-plugin/src/plugin/render.js')],
  });

  config.module
    .rule('jsx')
    .test(/\.(js|mjs|jsx)$/)
    .use('babel-loader')
    .loader(require.resolve('babel-loader'))
    .options(mergeBabelConfig);

  config.module
    .rule('tsx')
    .test(/\.tsx?$/)
    .use('string-replace')
    .loader(require.resolve('string-replace-loader'))
    .options({
      search: '__VERSION__',
      replace: pkg.version || 'null',
    })
    .end()
    .use('babel-loader')
    .loader(require.resolve('babel-loader'))
    .options(mergeBabelConfig)
    .end()
    .use('ts')
    .loader(require.resolve('ts-loader'));

  config.module
    .rule('md')
    .test(/\.md$/)
    .use('babel-loader')
    .loader(require.resolve('babel-loader'))
    .options(babelConfig)
    .end()
    .use('markdown-loader')
    .loader(require.resolve('../../webpackLoader/raxDemoLoader'));

  config.module
    .rule('assets')
    .test(/\.(svg|png|webp|jpe?g|gif)$/i)
    .use('source')
    .loader(require.resolve('image-source-loader'));

  config.module
    .rule('rxml')
    .test(/\.(rxml|xml)$/)
    .use('babel-loader')
    .loader(require.resolve('babel-loader'))
    .options({
      ...babelConfig,
      plugins: babelConfig.plugins.concat([[require.resolve('cloud-retail-rax-plugin/src/plugin/safety.js')]]),
    })
    .end()
    .use('rxml')
    .loader(require.resolve('cloud-retail-rax-plugin/src/plugin/index.js'))
    .options({})
    .end();

  config.plugin('caseSensitivePaths').use(CaseSensitivePathsPlugin);

  config.plugin('noError').use(webpack.NoEmitOnErrorsPlugin);

  if (enableProgress) {
    config.plugin('ProgressPlugin').use(ProgressPlugin, [
      {
        color: progressBarColor,
        name: name || target || 'webpack',
      },
    ]);
  }

  if (!options.minify) {
    // disable minify code
    config.optimization.minimize(options.minify);
  }

  if (command === 'start') {
    config.mode('development');
    config.devtool('inline-module-source-map');
  } else {
    config.mode('production');
    // support production sourcemap
    if (options.sourceMap) {
      config.devtool(typeof options.sourceMap === 'string' ? options.sourceMap : 'source-map');
    }
    config.optimization
      .minimizer('terser')
      .use(TerserPlugin, [
        {
          terserOptions: {
            output: {
              ascii_only: true,
              comments: false,
            },
          },
          extractComments: false,
          sourceMap: !!options.sourceMap,
        },
      ])
      .end()
      .minimizer('optimizeCSS')
      .use(OptimizeCSSAssetsPlugin);
  }

  if (pkg.name) {
    config.resolve.alias.set(pkg.name, path.resolve(rootDir, 'src/index'));
  }

  return config;
};
