// whale-component-docgen/src/render/renderComponentDoc.ts
import { WmComponentDoc, isWmUnionPropItemType } from 'whale-docgen-ts';
import {
  componentDocsMdastBuilder,
  tableMdastBuilder,
} from 'react-docgen-typescript-markdown-render';

import { heading, text, inlineCode, html, strong } from 'mdast-builder';
// import * as y18n from 'y18n';
// import * as path from 'path';

export const renderComponentDocs = (docs: WmComponentDoc[]) =>
  componentDocsMdastBuilder(docs, aliMaterialRenderer, {
    language: 'zh_CN',
  });

/**
 * alibaba materials style renderer
 * @see https://yuque.antfin-inc.com/mo/spec/spec-materials#CZiWi
 * @example
 * ``` md
 * ### Column
 *
 * Form column.
 *
 * | Param              | Description              | Type                                | Enum                            | Default |
 * | ------------------ | ------------------------ | ----------------------------------- | ------------------------------- | ------- |
 * | prop1              | prop1 description        | string                              |                                 | "red"   |
 * | prop2 _(required)_ | prop2 description        | number                              |                                 |         |
 * | prop3 _(required)_ | prop3 description a \| b | () => void                          |                                 |         |
 * | prop4 _(required)_ | prop4 description 中文   | "option1" \| "option2" \| "option3" | "option1", "option2", "option3" |         |
 * ```
 */
export const aliMaterialRenderer = (
  componentDoc: WmComponentDoc
  // { language }
) => {
  // console.log('language', language);

  const __ = (name: string) => name;
  // const { __ } = y18n({
  //   locale: language || 'en_US',
  //   directory: path.resolve(
  //     __dirname,
  //     '..',
  //     '..',
  //     'locales',
  //     'aliMaterialRenderer'
  //   ),
  // });
  return [
    heading(3, text(componentDoc.displayName)),
    text(componentDoc.description),
    tableMdastBuilder(Object.values(componentDoc.props), [
      {
        title: __('属性'),
        render: (vo) => text(vo.name),
      },
      {
        title: __('描述'),
        render: (vo) => [
          text(vo.description),
          ...(isWmUnionPropItemType(vo.type)
            ? [
                html('<br/>'),
                html('<br/>'),
                strong(text('可选值')),
                html('<br/>'),
                text(vo.type.value.map((e) => e.raw).join(', ')),
              ]
            : []),
        ],
      },
      {
        title: __('必填'),
        render: (vo) => (vo.required ? '✅' : ''),
      },
      {
        title: __('类型'),
        render: (vo) =>
          inlineCode(
            isWmUnionPropItemType(vo.type)
              ? 'Enum'
              : (vo.type.raw || '').replace(/\|/g, '&#124;')
          ),
      },
      {
        title: __('默认值'),
        render: (vo) => vo.defaultValue?.value,
      },
    ]),
  ];
};
