// build-plugin-component-multiple/src/template/home/index.js
import { createElement, Fragment } from 'rax';
import Image from 'rax-image';
import RaxIcon from 'rax-icon';
import { Page, Card } from '@alifd/meet-demo-layout';
import List from '@alifd/meet/es/list';
import ActionSheet from '@alifd/meet/es/action-sheet';
import Typography from '@alifd/meet/es/typography';
import Tag from '@alifd/meet/es/tag';
import navigate from '@uni/navigate';
import data from './data.json';
import '@alifd/meet/es/core/index.css';
import './index.css';

// generated by hbs
const LOGO_SOURCE = {
  uri: `{{logo}}`,
};
const VERSION = `{{version}}`;
const DESCRIPTION = `{{description}}`;

export default () => {
  const handleClick = (demos) => {
    if (!Array.isArray(demos)) {
      return;
    }

    if (demos.length === 1) {
      navigate
        .push({
          url: demos[0].miniappPageHref,
        })
        .then(() => {});
    } else {
      ActionSheet.show({
        message: '请选择要打开的 Demo',
        fullWidth: false,
        options: demos.map((demo) => demo.title),
        onClick: (option, index) => {
          navigate
            .push({
              url: demos[index].miniappPageHref,
            })
            .then(() => {});
        },
      });
    }
  };

  return (
    <Fragment>
      <Page fullWidth className="mp-page">
        <Card>
          <div className="mp-header">
            <Image source={LOGO_SOURCE} className="mp-logo" mode="widthFix" />
            <p>
              <Tag size="small">v{VERSION}</Tag>
            </p>
            <Typography type="body-1">{DESCRIPTION}</Typography>
          </div>
        </Card>
        <div className="mp-body">
          {data.map((group) => {
            const { components, name: groupName } = group;

            return (
              <Card key={groupName} fullWidth title={groupName}>
                <List>
                  {components.map((component) => {
                    const { componentName, componentPascalName, title: componentTitle, demos, icon } = component;
                    let Icon = null;

                    if (/:\/\//.test(icon)) {
                      const iconSource = { uri: icon };
                      Icon = <Image className="mp-img-icon" source={iconSource} />;
                    } else {
                      Icon = <RaxIcon className={`mp-thumb-icon mp-${componentName}`} />;
                    }

                    return (
                      <List.Item
                        media={Icon}
                        key={componentName}
                        title={componentTitle}
                        extra={componentPascalName}
                        arrow="right"
                        onClick={() => handleClick(demos)}
                      />
                    );
                  })}
                </List>
              </Card>
            );
          })}
        </div>
      </Page>
    </Fragment>
  );
};
