// @alifd/meet/lib/switch-field/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var view_1 = tslib_1.__importDefault(require("../view"));

var icon_1 = tslib_1.__importDefault(require("../icon"));

var field_1 = require("../field");

var switch_1 = tslib_1.__importDefault(require("../switch"));

require("./index.css");

var SwitchField = function (_super) {
  tslib_1.__extends(SwitchField, _super);

  function SwitchField() {
    return _super !== null && _super.apply(this, arguments) || this;
  }

  SwitchField.prototype.getValueName = function () {
    return 'checked';
  };

  SwitchField.prototype.renderPreview = function (value) {
    var _a = this.props.prefix,
        prefix = _a === void 0 ? 'mt-' : _a;
    return rax_1.createElement(view_1.default, {
      className: prefix + "switch-field-preview"
    }, rax_1.createElement(icon_1.default, {
      name: value ? 'select' : 'close'
    }));
  };

  SwitchField.prototype.renderControl = function () {
    return rax_1.createElement(switch_1.default, tslib_1.__assign({}, this.getProps()));
  };

  SwitchField.displayName = 'SwitchField';
  return SwitchField;
}(field_1.FieldElement);

exports.default = SwitchField;
