// @alifd/meet-react/es/upload/upload.js
import { rpx2vw4style as __rpx2vw__ } from "babel-runtime-jsx-style-transform";

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function _regeneratorRuntime() { return exports; }; var exports = {}, Op = Object.prototype, hasOwn = Op.hasOwnProperty, $Symbol = "function" == typeof Symbol ? Symbol : {}, iteratorSymbol = $Symbol.iterator || "@@iterator", asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator", toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag"; function define(obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: !0, configurable: !0, writable: !0 }), obj[key]; } try { define({}, ""); } catch (err) { define = function define(obj, key, value) { return obj[key] = value; }; } function wrap(innerFn, outerFn, self, tryLocsList) { var protoGenerator = outerFn && outerFn.prototype instanceof Generator ? outerFn : Generator, generator = Object.create(protoGenerator.prototype), context = new Context(tryLocsList || []); return generator._invoke = function (innerFn, self, context) { var state = "suspendedStart"; return function (method, arg) { if ("executing" === state) throw new Error("Generator is already running"); if ("completed" === state) { if ("throw" === method) throw arg; return doneResult(); } for (context.method = method, context.arg = arg;;) { var delegate = context.delegate; if (delegate) { var delegateResult = maybeInvokeDelegate(delegate, context); if (delegateResult) { if (delegateResult === ContinueSentinel) continue; return delegateResult; } } if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) { if ("suspendedStart" === state) throw state = "completed", context.arg; context.dispatchException(context.arg); } else "return" === context.method && context.abrupt("return", context.arg); state = "executing"; var record = tryCatch(innerFn, self, context); if ("normal" === record.type) { if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue; return { value: record.arg, done: context.done }; } "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg); } }; }(innerFn, self, context), generator; } function tryCatch(fn, obj, arg) { try { return { type: "normal", arg: fn.call(obj, arg) }; } catch (err) { return { type: "throw", arg: err }; } } exports.wrap = wrap; var ContinueSentinel = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var IteratorPrototype = {}; define(IteratorPrototype, iteratorSymbol, function () { return this; }); var getProto = Object.getPrototypeOf, NativeIteratorPrototype = getProto && getProto(getProto(values([]))); NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype); var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype); function defineIteratorMethods(prototype) { ["next", "throw", "return"].forEach(function (method) { define(prototype, method, function (arg) { return this._invoke(method, arg); }); }); } function AsyncIterator(generator, PromiseImpl) { function invoke(method, arg, resolve, reject) { var record = tryCatch(generator[method], generator, arg); if ("throw" !== record.type) { var result = record.arg, value = result.value; return value && "object" == _typeof(value) && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) { invoke("next", value, resolve, reject); }, function (err) { invoke("throw", err, resolve, reject); }) : PromiseImpl.resolve(value).then(function (unwrapped) { result.value = unwrapped, resolve(result); }, function (error) { return invoke("throw", error, resolve, reject); }); } reject(record.arg); } var previousPromise; this._invoke = function (method, arg) { function callInvokeWithMethodAndArg() { return new PromiseImpl(function (resolve, reject) { invoke(method, arg, resolve, reject); }); } return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); }; } function maybeInvokeDelegate(delegate, context) { var method = delegate.iterator[context.method]; if (undefined === method) { if (context.delegate = null, "throw" === context.method) { if (delegate.iterator["return"] && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method)) return ContinueSentinel; context.method = "throw", context.arg = new TypeError("The iterator does not provide a 'throw' method"); } return ContinueSentinel; } var record = tryCatch(method, delegate.iterator, context.arg); if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel; var info = record.arg; return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel); } function pushTryEntry(locs) { var entry = { tryLoc: locs[0] }; 1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry); } function resetTryEntry(entry) { var record = entry.completion || {}; record.type = "normal", delete record.arg, entry.completion = record; } function Context(tryLocsList) { this.tryEntries = [{ tryLoc: "root" }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0); } function values(iterable) { if (iterable) { var iteratorMethod = iterable[iteratorSymbol]; if (iteratorMethod) return iteratorMethod.call(iterable); if ("function" == typeof iterable.next) return iterable; if (!isNaN(iterable.length)) { var i = -1, next = function next() { for (; ++i < iterable.length;) { if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next; } return next.value = undefined, next.done = !0, next; }; return next.next = next; } } return { next: doneResult }; } function doneResult() { return { value: undefined, done: !0 }; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, define(Gp, "constructor", GeneratorFunctionPrototype), define(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) { var ctor = "function" == typeof genFun && genFun.constructor; return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name)); }, exports.mark = function (genFun) { return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun; }, exports.awrap = function (arg) { return { __await: arg }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () { return this; }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) { void 0 === PromiseImpl && (PromiseImpl = Promise); var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl); return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) { return result.done ? result.value : iter.next(); }); }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () { return this; }), define(Gp, "toString", function () { return "[object Generator]"; }), exports.keys = function (object) { var keys = []; for (var key in object) { keys.push(key); } return keys.reverse(), function next() { for (; keys.length;) { var key = keys.pop(); if (key in object) return next.value = key, next.done = !1, next; } return next.done = !0, next; }; }, exports.values = values, Context.prototype = { constructor: Context, reset: function reset(skipTempReset) { if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) { "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined); } }, stop: function stop() { this.done = !0; var rootRecord = this.tryEntries[0].completion; if ("throw" === rootRecord.type) throw rootRecord.arg; return this.rval; }, dispatchException: function dispatchException(exception) { if (this.done) throw exception; var context = this; function handle(loc, caught) { return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught; } for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i], record = entry.completion; if ("root" === entry.tryLoc) return handle("end"); if (entry.tryLoc <= this.prev) { var hasCatch = hasOwn.call(entry, "catchLoc"), hasFinally = hasOwn.call(entry, "finallyLoc"); if (hasCatch && hasFinally) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } else if (hasCatch) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); } else { if (!hasFinally) throw new Error("try statement without catch or finally"); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } } } }, abrupt: function abrupt(type, arg) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) { var finallyEntry = entry; break; } } finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null); var record = finallyEntry ? finallyEntry.completion : {}; return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record); }, complete: function complete(record, afterLoc) { if ("throw" === record.type) throw record.arg; return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel; }, finish: function finish(finallyLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel; } }, "catch": function _catch(tryLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc === tryLoc) { var record = entry.completion; if ("throw" === record.type) { var thrown = record.arg; resetTryEntry(entry); } return thrown; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(iterable, resultName, nextLoc) { return this.delegate = { iterator: values(iterable), resultName: resultName, nextLoc: nextLoc }, "next" === this.method && (this.arg = undefined), ContinueSentinel; } }, exports; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

import { __awaiter, __rest } from "tslib";
import React, { createElement, forwardRef, useCallback, useEffect, useImperativeHandle, useRef, useState } from "react";
import classNames from 'classnames';
import { Text } from "@alifd/meet-react-component-one";
import chooseImage from 'universal-choose-image';
import View from '../view';
import Icon from '../icon';
import Message from '../message';
import Card from './item/card';
import { useLocale } from '../locale';
import defaultLang from '../locale/lang/zh-cn';
import { default as defaultRequest } from './request';
import { getWidth, isBoolean, isFunction, isStrictWechatMiniProgram, isString, isValidArray, replaceMessage } from '../utils';
import { useGuid } from '../utils/hooks';
import { chooseImageForWechat, createFileItem, UPLOAD_THROTTLE } from './utils';
import { isPromise } from '../utils/types';
var FILE_CACHAE = {};

var updateFileStatus = function updateFileStatus() {
  var files = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];

  if (!isValidArray(files)) {
    return [];
  }

  return _toConsumableArray(files).map(function (file) {
    var uid = file.uid,
        state = file.state;

    if (FILE_CACHAE[uid] === 'done' && state === 'uploading') {
      file.state = 'done';
    }

    return file;
  });
};

var Upload = function Upload(props, ref) {
  var _props$prefix = props.prefix,
      prefix = _props$prefix === void 0 ? 'mt-' : _props$prefix,
      extraIcon = props.extraIcon,
      className = props.className,
      action = props.action,
      _props$limit = props.limit,
      limit = _props$limit === void 0 ? 1 : _props$limit,
      _props$label = props.label,
      label = _props$label === void 0 ? '' : _props$label,
      valueProp = props.value,
      defaultValue = props.defaultValue,
      dataProp = props.data,
      headers = props.headers,
      request = props.request,
      formatter = props.formatter,
      disabled = props.disabled,
      _props$cols = props.cols,
      cols = _props$cols === void 0 ? 4 : _props$cols,
      _props$size = props.size,
      size = _props$size === void 0 ? 'medium' : _props$size,
      _props$listType = props.listType,
      listType = _props$listType === void 0 ? 'card' : _props$listType,
      _props$name = props.name,
      nameProp = _props$name === void 0 ? 'file' : _props$name,
      filePicker = props.filePicker,
      withCredentials = props.withCredentials,
      fileKeyName = props.fileKeyName,
      _props$fileType = props.fileType,
      fileType = _props$fileType === void 0 ? 'image' : _props$fileType,
      _props$accept = props.accept,
      accept = _props$accept === void 0 ? 'image/*' : _props$accept,
      _props$useDataURL = props.useDataURL,
      useDataURL = _props$useDataURL === void 0 ? true : _props$useDataURL,
      _props$autoUpload = props.autoUpload,
      autoUpload = _props$autoUpload === void 0 ? true : _props$autoUpload,
      _props$hasRemove = props.hasRemove,
      hasRemove = _props$hasRemove === void 0 ? true : _props$hasRemove,
      _props$showAddButton = props.showAddButton,
      showAddButton = _props$showAddButton === void 0 ? true : _props$showAddButton,
      _props$singleLimit = props.singleLimit,
      singleLimit = _props$singleLimit === void 0 ? 9 : _props$singleLimit,
      beforeUpload = props.beforeUpload,
      afterSelect = props.afterSelect,
      onProgress = props.onProgress,
      _onSuccess = props.onSuccess,
      _onError = props.onError,
      progressProps = props.progressProps,
      onSelect = props.onSelect,
      onChange = props.onChange,
      onRemove = props.onRemove,
      onPreview = props.onPreview,
      _props$sizeType = props.sizeType,
      sizeType = _props$sizeType === void 0 ? ['original', 'compressed'] : _props$sizeType,
      _props$sourceType = props.sourceType,
      sourceType = _props$sourceType === void 0 ? ['album', 'camera'] : _props$sourceType,
      children = props.children,
      style = props.style,
      others = __rest(props, ["prefix", "extraIcon", "className", "action", "limit", "label", "value", "defaultValue", "data", "headers", "request", "formatter", "disabled", "cols", "size", "listType", "name", "filePicker", "withCredentials", "fileKeyName", "fileType", "accept", "useDataURL", "autoUpload", "hasRemove", "showAddButton", "singleLimit", "beforeUpload", "afterSelect", "onProgress", "onSuccess", "onError", "progressProps", "onSelect", "onChange", "onRemove", "onPreview", "sizeType", "sourceType", "children", "style"]);

  var name = fileKeyName || nameProp;
  var locale = useLocale('Upload', others, defaultLang.Upload);

  var _useState = useState('0px'),
      _useState2 = _slicedToArray(_useState, 2),
      itemSize = _useState2[0],
      setItemSize = _useState2[1];

  var _useState3 = useState(valueProp || defaultValue || []),
      _useState4 = _slicedToArray(_useState3, 2),
      value = _useState4[0],
      setValue = _useState4[1];

  var newestValue = useRef(value);
  var timerQueue = useRef([]);
  var uploadRef = useRef();
  var uploadId = useGuid('mtUpload');
  var uploadStyle = Object.assign({
    gridTemplateColumns: "repeat(".concat(cols, ", 1fr)")
  }, style);
  useEffect(function () {
    if (isValidArray(valueProp)) {
      valueProp.forEach(function (file) {
        if (file.state && file.uid) {
          FILE_CACHAE[file.uid] = file.state;
        }
      });
    }

    setTimeout(function () {
      (function () {
        return __awaiter(void 0, void 0, void 0, /*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
          var width, newCardSize;
          return _regeneratorRuntime().wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  _context.next = 2;
                  return getWidth(uploadId);

                case 2:
                  width = _context.sent;
                  newCardSize = "calc(((".concat(width, "px - var(--upload-card-spacing)*").concat(cols - 1, " )) / ").concat(cols, ")");
                  setItemSize(newCardSize);

                case 5:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee);
        }));
      })();
    }, 200);
    return function () {
      timerQueue.current.forEach(function (x) {
        return clearInterval(x);
      });
      timerQueue.current = [];
    };
  }, []);
  var getFile = useCallback(function (uid) {
    var item = newestValue.current.find(function (it) {
      return it.uid === uid;
    });

    if (item) {
      return {
        file: item,
        index: newestValue.current.indexOf(item)
      };
    }

    return undefined;
  }, []);
  var updateFile = useCallback(function (uid, mod) {
    var gotFile = getFile(uid);

    if (!gotFile) {
      throw new Error("File id ".concat(uid, " not found"));
    }

    var file = gotFile.file,
        index = gotFile.index;

    if (mod.state) {
      FILE_CACHAE[uid] = mod.state;
    }

    var isSame = true;

    for (var k in mod) {
      if (mod.hasOwnProperty(k) && file[k] !== mod[k]) {
        isSame = false;
        break;
      }
    }

    if (isSame) {
      return file;
    }

    var newFile = Object.assign(Object.assign({}, file), mod);

    var newValue = _toConsumableArray(newestValue.current);

    newValue.splice(index, 1, newFile);
    newestValue.current = newValue;

    if (onChange) {
      onChange(newValue);
    }

    if (!('value' in props)) {
      setValue(newValue);
    }

    return newFile;
  }, [valueProp, onChange]);
  var uploadFiles = useCallback(function (files) {
    var uploadFile = function uploadFile(it) {
      return __awaiter(void 0, void 0, void 0, /*#__PURE__*/_regeneratorRuntime().mark(function _callee2() {
        var uid, options, res, curFile, timer, clearTimer, complete, newProcess;
        return _regeneratorRuntime().wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (!(it.state !== 'selected')) {
                  _context2.next = 2;
                  break;
                }

                return _context2.abrupt("return");

              case 2:
                uid = it.uid;
                updateFile(uid, {
                  state: 'pending'
                });
                options = {
                  action: action,
                  data: dataProp || {},
                  headers: headers || {}
                };

                if (isFunction(options.data)) {
                  options.data = options.data();
                }

                if (!beforeUpload) {
                  _context2.next = 29;
                  break;
                }

                _context2.prev = 7;
                res = beforeUpload(it, options);

                if (!isBoolean(res)) {
                  _context2.next = 15;
                  break;
                }

                if (!(res === false)) {
                  _context2.next = 13;
                  break;
                }

                updateFile(uid, {
                  state: 'uploading'
                });
                return _context2.abrupt("return");

              case 13:
                _context2.next = 22;
                break;

              case 15:
                if (!isPromise(res)) {
                  _context2.next = 21;
                  break;
                }

                _context2.next = 18;
                return res;

              case 18:
                options = _context2.sent;
                _context2.next = 22;
                break;

              case 21:
                options = res;

              case 22:
                _context2.next = 29;
                break;

              case 24:
                _context2.prev = 24;
                _context2.t0 = _context2["catch"](7);
                updateFile(uid, {
                  state: 'error'
                });
                console.error(_context2.t0);
                return _context2.abrupt("return");

              case 29:
                curFile = updateFile(uid, {
                  state: 'uploading'
                });

                if (onProgress) {
                  onProgress();
                }

                if (options === null || options === void 0 ? void 0 : options.action) {
                  _context2.next = 36;
                  break;
                }

                console.warn('Upload request failed,  action prop is required');
                curFile = updateFile(uid, {
                  state: 'error'
                });

                if (_onError) {
                  _onError(curFile, newestValue.current);
                }

                return _context2.abrupt("return");

              case 36:
                timer = null;

                clearTimer = function clearTimer() {
                  if (timer && timerQueue.current.includes(timer)) {
                    clearInterval(timer);
                    timerQueue.current.splice(timerQueue.current.indexOf(timer), 1);
                  }

                  timer = null;
                };

                complete = function complete() {
                  return clearTimer();
                };

                newProcess = 0;
                _context2.next = 42;
                return new Promise(function (resolve, reject) {
                  var requestObj = {
                    onProgress: function onProgress(ev) {
                      if (curFile.state !== 'uploading') {
                        return;
                      }

                      if (!timer) {
                        timer = setInterval(function () {
                          if (curFile.state !== 'uploading') {
                            clearTimer();
                            return;
                          }

                          curFile = updateFile(uid, {
                            state: 'uploading',
                            percent: newProcess
                          });

                          if (newProcess === 100) {
                            clearTimer();
                          }
                        }, UPLOAD_THROTTLE);
                        timerQueue.current.push(timer);
                      }

                      newProcess = ev.percent;
                    },
                    onError: function onError(err, body) {
                      console.error('Upload request failed', err);
                      curFile = updateFile(uid, {
                        state: 'error',
                        response: body
                      });

                      if (_onError) {
                        _onError(curFile, newestValue.current);
                      }

                      resolve('error');
                      complete();
                    },
                    onSuccess: function onSuccess(res) {
                      res.status = res.statusCode;

                      if (isString(res.data)) {
                        try {
                          res.data = JSON.parse(res.data);
                        } catch (e) {}
                      }

                      var result = formatter ? formatter(res, it) : res.data;

                      if (_typeof(result) !== 'object') {
                        console.error("Expect a json object response, ".concat(result, " got"));
                        curFile = updateFile(uid, {
                          state: 'error'
                        });

                        if (_onError) {
                          _onError(curFile, newestValue.current);
                        }

                        resolve('error');
                        return;
                      }

                      var finalState = {
                        response: result,
                        percent: 100,
                        state: result.success ? 'done' : 'error'
                      };

                      if (result.success && result.url) {
                        finalState.url = result.url;
                        finalState.imgURL = result.imgURL || result.url;
                        finalState.downloadURL = result.downloadURL || result.url;
                      }

                      curFile = updateFile(uid, finalState);

                      if (finalState.state === 'done' && _onSuccess) {
                        _onSuccess(curFile, newestValue.current);
                      }

                      if (finalState.state === 'error' && _onError) {
                        _onError(curFile, newestValue.current);
                      }

                      clearTimer();
                      resolve('success');
                    },
                    fileType: fileType,
                    data: options.data,
                    filename: name,
                    file: it.file,
                    withCredentials: withCredentials,
                    action: options.action,
                    method: 'post',
                    timeout: 0,
                    headers: options.headers
                  };

                  if (request) {
                    request(requestObj);
                  } else {
                    defaultRequest(requestObj);
                  }
                });

              case 42:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, null, [[7, 24]]);
      }));
    };

    if (!isValidArray(files)) {
      return;
    }

    try {
      Promise.all(files.map(function (it) {
        return uploadFile(it);
      })).then(function () {
        var arr = updateFileStatus(newestValue.current);

        if (arr) {
          if (isFunction(onChange)) {
            onChange(arr);
          }

          if (!('value' in props)) {
            setValue(arr);
            newestValue.current = arr;
          }
        }
      });
    } catch (e) {
      console.warn('上传失败');
    }
  }, [action, dataProp, headers, beforeUpload, onProgress, formatter, _onSuccess, _onError, request]);
  var doUpload = useCallback(function () {
    return uploadFiles(newestValue.current.filter(function (it) {
      return it.state === 'selected';
    }));
  }, []);
  var doReUpload = useCallback(function (it) {
    updateFile(it.uid, {
      state: 'selected',
      percent: 0
    });
    doUpload();
  }, []);
  var handleAdd = useCallback(function () {
    return __awaiter(void 0, void 0, void 0, /*#__PURE__*/_regeneratorRuntime().mark(function _callee3() {
      var file, defaultSizeType, defaultSourceType, mediaType, fileItems, newValue;
      return _regeneratorRuntime().wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              if (!(limit <= newestValue.current.length || disabled)) {
                _context3.next = 3;
                break;
              }

              Message.notice(replaceMessage(locale.limit, {
                limit: limit
              }));
              return _context3.abrupt("return");

            case 3:
              defaultSizeType = ['original', 'compressed'];
              defaultSourceType = ['album', 'camera'];
              _context3.prev = 5;

              if (!isFunction(filePicker)) {
                _context3.next = 12;
                break;
              }

              _context3.next = 9;
              return filePicker();

            case 9:
              file = _context3.sent;
              _context3.next = 23;
              break;

            case 12:
              if (!isStrictWechatMiniProgram) {
                _context3.next = 20;
                break;
              }

              mediaType = ['video', 'image'];

              if (accept === 'image/*') {
                mediaType = ['image'];
              } else if (accept === 'video/*') {
                mediaType = ['video'];
              }

              _context3.next = 17;
              return chooseImageForWechat({
                count: Math.min(singleLimit, limit - newestValue.current.length),
                mediaType: mediaType,
                sizeType: sizeType || defaultSizeType,
                sourceType: sourceType || defaultSourceType
              });

            case 17:
              file = _context3.sent;
              _context3.next = 23;
              break;

            case 20:
              _context3.next = 22;
              return chooseImage({
                count: Math.min(singleLimit, limit - newestValue.current.length),
                accept: accept,
                sizeType: sizeType || defaultSizeType,
                sourceType: sourceType || defaultSourceType
              });

            case 22:
              file = _context3.sent;

            case 23:
              _context3.next = 28;
              break;

            case 25:
              _context3.prev = 25;
              _context3.t0 = _context3["catch"](5);
              console.error(_context3.t0);

            case 28:
              if (!(!file || !Array.isArray(file === null || file === void 0 ? void 0 : file.data))) {
                _context3.next = 30;
                break;
              }

              return _context3.abrupt("return");

            case 30:
              fileItems = file.data.map(function (it, index) {
                return createFileItem(it, file.files ? file.files[index] : null);
              });

              if (afterSelect) {
                fileItems = fileItems.filter(function (it) {
                  return afterSelect(it) !== false;
                });
              }

              if (useDataURL) {
                fileItems.forEach(function (it) {
                  if (it.dataURL && it.isImage) {
                    it.imgURL = it.dataURL;
                  }
                });
              }

              if (onSelect) {
                onSelect();
              }

              newValue = [].concat(_toConsumableArray(newestValue.current), _toConsumableArray(fileItems));

              if (onChange) {
                onChange(newValue);
              }

              if (!('value' in props)) {
                setValue(newValue);
                newestValue.current = newValue;
              }

              if (autoUpload) {
                doUpload();
              }

            case 38:
            case "end":
              return _context3.stop();
          }
        }
      }, _callee3, null, [[5, 25]]);
    }));
  }, [limit, disabled, locale, sizeType, sourceType, afterSelect, autoUpload, onSelect, onChange]);
  var handleRemove = useCallback(function (file) {
    var doRemove = function doRemove() {
      var index = newestValue.current.indexOf(file);

      if (index === -1) {
        return;
      }

      var newValue = _toConsumableArray(newestValue.current);

      newValue.splice(index, 1);

      if (onChange) {
        onChange(newValue);
      }

      if (!('value' in props)) {
        setValue(newValue);
        newestValue.current = newValue;
      }
    };

    if (!onRemove) {
      doRemove();
      return;
    }

    var result = onRemove(file);

    if (isBoolean(result)) {
      if (result === true) {
        doRemove();
      }

      return;
    }

    if (isPromise(result)) {
      result.then(function (should) {
        return should && doRemove();
      })["catch"](function () {});
    }
  }, [onChange, onRemove]);
  var handleItemClick = useCallback(function (file, e) {
    if (onPreview) {
      onPreview(file, Object.assign(Object.assign({}, e), {
        fileList: value
      }));
    }
  }, [onPreview]);
  useEffect(function () {
    if ('value' in props) {
      newestValue.current = updateFileStatus(valueProp || defaultValue || []);
      setValue(newestValue.current);

      if (autoUpload) {
        doUpload();
      }
    }
  }, [valueProp]);
  useImperativeHandle(ref, function () {
    return Object.assign(uploadRef.current, {
      add: handleAdd,
      upload: doUpload,
      getInstance: function getInstance() {
        return {
          startUpload: doUpload
        };
      }
    });
  });
  var clsPrefix = "".concat(prefix, "upload");
  var hasAdd = showAddButton && limit > value.length && !disabled;
  var ItemComponent = Card;
  return /*#__PURE__*/React.createElement(View, {
    id: uploadId,
    ref: uploadRef
  }, label || extraIcon ? /*#__PURE__*/React.createElement(View, {
    className: "".concat(clsPrefix, "-header")
  }, /*#__PURE__*/React.createElement(Text, {
    className: "".concat(clsPrefix, "-label ").concat(clsPrefix, "-label--").concat(size)
  }, label), extraIcon ? /*#__PURE__*/React.createElement(View, {
    className: "".concat(clsPrefix, "-extra")
  }, /*#__PURE__*/React.createElement(Text, {
    className: "".concat(clsPrefix, "-counter")
  }, value.length || 0, "/", limit), /*#__PURE__*/React.createElement(Icon, {
    className: "".concat(clsPrefix, "-extra-icon"),
    size: size,
    onClick: handleAdd,
    name: extraIcon
  })) : null) : null, /*#__PURE__*/React.createElement(View, _extends({
    className: classNames(clsPrefix, "".concat(clsPrefix, "--").concat(listType), className),
    style: __rpx2vw__(uploadStyle)
  }, others), value.map(function (file) {
    return /*#__PURE__*/React.createElement(ItemComponent, {
      key: file.uid,
      file: file,
      prefix: clsPrefix,
      progressProps: progressProps,
      onClick: handleItemClick,
      reUpload: locale.reupload,
      onReUpload: doReUpload,
      hasRemove: hasRemove && !disabled,
      onRemove: handleRemove,
      style: __rpx2vw__({
        width: "".concat(itemSize),
        height: "".concat(itemSize)
      })
    });
  }), hasAdd && /*#__PURE__*/React.createElement(View, {
    className: classNames("".concat(clsPrefix, "-item"), "".concat(clsPrefix, "-item--add"), "".concat(clsPrefix, "-item--").concat(listType)),
    onClick: handleAdd,
    style: __rpx2vw__({
      width: "".concat(itemSize),
      height: "".concat(itemSize)
    })
  }, children ? children : /*#__PURE__*/React.createElement(Icon, {
    className: "".concat(clsPrefix, "-item-icon"),
    name: "add"
  }))));
};

Upload.displayName = 'Upload';
export default /*#__PURE__*/forwardRef(Upload);
