// whale-docgen-ts/lib/utils/propItemType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWmPlainPropItemType = exports.isWmArrayPropItemType = exports.isWmObjectPropItemType = exports.isWmIntersectionPropItemType = exports.isWmUnionPropItemType = exports.isWmPlainPropItemType = void 0;
const isWmPlainPropItemType = (type) => type.name === 'plain';
exports.isWmPlainPropItemType = isWmPlainPropItemType;
const isWmUnionPropItemType = (type) => type.name === 'union';
exports.isWmUnionPropItemType = isWmUnionPropItemType;
const isWmIntersectionPropItemType = (type) => type.name === 'intersection';
exports.isWmIntersectionPropItemType = isWmIntersectionPropItemType;
const isWmObjectPropItemType = (type) => type.name === 'object';
exports.isWmObjectPropItemType = isWmObjectPropItemType;
const isWmArrayPropItemType = (type) => type.name === 'array';
exports.isWmArrayPropItemType = isWmArrayPropItemType;
const createWmPlainPropItemType = (raw) => ({
    name: 'plain',
    raw,
});
exports.createWmPlainPropItemType = createWmPlainPropItemType;
