// @alifd/meet-react/lib/upload/upload.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _babelRuntimeJsxStyleTransform = require("babel-runtime-jsx-style-transform");

var _tslib = require("tslib");

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _meetReactComponentOne = require("@alifd/meet-react-component-one");

var _universalChooseImage = _interopRequireDefault(require("universal-choose-image"));

var _view = _interopRequireDefault(require("../view"));

var _icon = _interopRequireDefault(require("../icon"));

var _message = _interopRequireDefault(require("../message"));

var _card = _interopRequireDefault(require("./item/card"));

var _locale = require("../locale");

var _zhCn = _interopRequireDefault(require("../locale/lang/zh-cn"));

var _request = _interopRequireDefault(require("./request"));

var _utils = require("../utils");

var _hooks = require("../utils/hooks");

var _utils2 = require("./utils");

var _types = require("../utils/types");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function () { return exports; }; var exports = {}, Op = Object.prototype, hasOwn = Op.hasOwnProperty, $Symbol = "function" == typeof Symbol ? Symbol : {}, iteratorSymbol = $Symbol.iterator || "@@iterator", asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator", toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag"; function define(obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: !0, configurable: !0, writable: !0 }), obj[key]; } try { define({}, ""); } catch (err) { define = function (obj, key, value) { return obj[key] = value; }; } function wrap(innerFn, outerFn, self, tryLocsList) { var protoGenerator = outerFn && outerFn.prototype instanceof Generator ? outerFn : Generator, generator = Object.create(protoGenerator.prototype), context = new Context(tryLocsList || []); return generator._invoke = function (innerFn, self, context) { var state = "suspendedStart"; return function (method, arg) { if ("executing" === state) throw new Error("Generator is already running"); if ("completed" === state) { if ("throw" === method) throw arg; return doneResult(); } for (context.method = method, context.arg = arg;;) { var delegate = context.delegate; if (delegate) { var delegateResult = maybeInvokeDelegate(delegate, context); if (delegateResult) { if (delegateResult === ContinueSentinel) continue; return delegateResult; } } if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) { if ("suspendedStart" === state) throw state = "completed", context.arg; context.dispatchException(context.arg); } else "return" === context.method && context.abrupt("return", context.arg); state = "executing"; var record = tryCatch(innerFn, self, context); if ("normal" === record.type) { if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue; return { value: record.arg, done: context.done }; } "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg); } }; }(innerFn, self, context), generator; } function tryCatch(fn, obj, arg) { try { return { type: "normal", arg: fn.call(obj, arg) }; } catch (err) { return { type: "throw", arg: err }; } } exports.wrap = wrap; var ContinueSentinel = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var IteratorPrototype = {}; define(IteratorPrototype, iteratorSymbol, function () { return this; }); var getProto = Object.getPrototypeOf, NativeIteratorPrototype = getProto && getProto(getProto(values([]))); NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype); var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype); function defineIteratorMethods(prototype) { ["next", "throw", "return"].forEach(function (method) { define(prototype, method, function (arg) { return this._invoke(method, arg); }); }); } function AsyncIterator(generator, PromiseImpl) { function invoke(method, arg, resolve, reject) { var record = tryCatch(generator[method], generator, arg); if ("throw" !== record.type) { var result = record.arg, value = result.value; return value && "object" == typeof value && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) { invoke("next", value, resolve, reject); }, function (err) { invoke("throw", err, resolve, reject); }) : PromiseImpl.resolve(value).then(function (unwrapped) { result.value = unwrapped, resolve(result); }, function (error) { return invoke("throw", error, resolve, reject); }); } reject(record.arg); } var previousPromise; this._invoke = function (method, arg) { function callInvokeWithMethodAndArg() { return new PromiseImpl(function (resolve, reject) { invoke(method, arg, resolve, reject); }); } return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); }; } function maybeInvokeDelegate(delegate, context) { var method = delegate.iterator[context.method]; if (undefined === method) { if (context.delegate = null, "throw" === context.method) { if (delegate.iterator.return && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method)) return ContinueSentinel; context.method = "throw", context.arg = new TypeError("The iterator does not provide a 'throw' method"); } return ContinueSentinel; } var record = tryCatch(method, delegate.iterator, context.arg); if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel; var info = record.arg; return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel); } function pushTryEntry(locs) { var entry = { tryLoc: locs[0] }; 1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry); } function resetTryEntry(entry) { var record = entry.completion || {}; record.type = "normal", delete record.arg, entry.completion = record; } function Context(tryLocsList) { this.tryEntries = [{ tryLoc: "root" }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0); } function values(iterable) { if (iterable) { var iteratorMethod = iterable[iteratorSymbol]; if (iteratorMethod) return iteratorMethod.call(iterable); if ("function" == typeof iterable.next) return iterable; if (!isNaN(iterable.length)) { var i = -1, next = function next() { for (; ++i < iterable.length;) { if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next; } return next.value = undefined, next.done = !0, next; }; return next.next = next; } } return { next: doneResult }; } function doneResult() { return { value: undefined, done: !0 }; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, define(Gp, "constructor", GeneratorFunctionPrototype), define(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) { var ctor = "function" == typeof genFun && genFun.constructor; return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name)); }, exports.mark = function (genFun) { return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun; }, exports.awrap = function (arg) { return { __await: arg }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () { return this; }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) { void 0 === PromiseImpl && (PromiseImpl = Promise); var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl); return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) { return result.done ? result.value : iter.next(); }); }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () { return this; }), define(Gp, "toString", function () { return "[object Generator]"; }), exports.keys = function (object) { var keys = []; for (var key in object) { keys.push(key); } return keys.reverse(), function next() { for (; keys.length;) { var key = keys.pop(); if (key in object) return next.value = key, next.done = !1, next; } return next.done = !0, next; }; }, exports.values = values, Context.prototype = { constructor: Context, reset: function reset(skipTempReset) { if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) { "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined); } }, stop: function stop() { this.done = !0; var rootRecord = this.tryEntries[0].completion; if ("throw" === rootRecord.type) throw rootRecord.arg; return this.rval; }, dispatchException: function dispatchException(exception) { if (this.done) throw exception; var context = this; function handle(loc, caught) { return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught; } for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i], record = entry.completion; if ("root" === entry.tryLoc) return handle("end"); if (entry.tryLoc <= this.prev) { var hasCatch = hasOwn.call(entry, "catchLoc"), hasFinally = hasOwn.call(entry, "finallyLoc"); if (hasCatch && hasFinally) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } else if (hasCatch) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); } else { if (!hasFinally) throw new Error("try statement without catch or finally"); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } } } }, abrupt: function abrupt(type, arg) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) { var finallyEntry = entry; break; } } finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null); var record = finallyEntry ? finallyEntry.completion : {}; return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record); }, complete: function complete(record, afterLoc) { if ("throw" === record.type) throw record.arg; return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel; }, finish: function finish(finallyLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel; } }, catch: function _catch(tryLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc === tryLoc) { var record = entry.completion; if ("throw" === record.type) { var thrown = record.arg; resetTryEntry(entry); } return thrown; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(iterable, resultName, nextLoc) { return this.delegate = { iterator: values(iterable), resultName: resultName, nextLoc: nextLoc }, "next" === this.method && (this.arg = undefined), ContinueSentinel; } }, exports; }

var FILE_CACHAE = {};

var updateFileStatus = function (files) {
  if (files === void 0) {
    files = [];
  }

  if (!(0, _utils.isValidArray)(files)) {
    return [];
  }

  return [].concat(files).map(function (file) {
    var uid = file.uid,
        state = file.state;

    if (FILE_CACHAE[uid] === 'done' && state === 'uploading') {
      file.state = 'done';
    }

    return file;
  });
};

var Upload = function (props, ref) {
  var _props$prefix = props.prefix,
      prefix = _props$prefix === void 0 ? 'mt-' : _props$prefix,
      extraIcon = props.extraIcon,
      className = props.className,
      action = props.action,
      _props$limit = props.limit,
      limit = _props$limit === void 0 ? 1 : _props$limit,
      _props$label = props.label,
      label = _props$label === void 0 ? '' : _props$label,
      valueProp = props.value,
      defaultValue = props.defaultValue,
      dataProp = props.data,
      headers = props.headers,
      request = props.request,
      formatter = props.formatter,
      disabled = props.disabled,
      _props$cols = props.cols,
      cols = _props$cols === void 0 ? 4 : _props$cols,
      _props$size = props.size,
      size = _props$size === void 0 ? 'medium' : _props$size,
      _props$listType = props.listType,
      listType = _props$listType === void 0 ? 'card' : _props$listType,
      _props$name = props.name,
      nameProp = _props$name === void 0 ? 'file' : _props$name,
      filePicker = props.filePicker,
      withCredentials = props.withCredentials,
      fileKeyName = props.fileKeyName,
      _props$fileType = props.fileType,
      fileType = _props$fileType === void 0 ? 'image' : _props$fileType,
      _props$accept = props.accept,
      accept = _props$accept === void 0 ? 'image/*' : _props$accept,
      _props$useDataURL = props.useDataURL,
      useDataURL = _props$useDataURL === void 0 ? true : _props$useDataURL,
      _props$autoUpload = props.autoUpload,
      autoUpload = _props$autoUpload === void 0 ? true : _props$autoUpload,
      _props$hasRemove = props.hasRemove,
      hasRemove = _props$hasRemove === void 0 ? true : _props$hasRemove,
      _props$showAddButton = props.showAddButton,
      showAddButton = _props$showAddButton === void 0 ? true : _props$showAddButton,
      _props$singleLimit = props.singleLimit,
      singleLimit = _props$singleLimit === void 0 ? 9 : _props$singleLimit,
      beforeUpload = props.beforeUpload,
      afterSelect = props.afterSelect,
      onProgress = props.onProgress,
      _onSuccess = props.onSuccess,
      _onError = props.onError,
      progressProps = props.progressProps,
      onSelect = props.onSelect,
      onChange = props.onChange,
      onRemove = props.onRemove,
      onPreview = props.onPreview,
      _props$sizeType = props.sizeType,
      sizeType = _props$sizeType === void 0 ? ['original', 'compressed'] : _props$sizeType,
      _props$sourceType = props.sourceType,
      sourceType = _props$sourceType === void 0 ? ['album', 'camera'] : _props$sourceType,
      children = props.children,
      style = props.style,
      others = (0, _tslib.__rest)(props, ["prefix", "extraIcon", "className", "action", "limit", "label", "value", "defaultValue", "data", "headers", "request", "formatter", "disabled", "cols", "size", "listType", "name", "filePicker", "withCredentials", "fileKeyName", "fileType", "accept", "useDataURL", "autoUpload", "hasRemove", "showAddButton", "singleLimit", "beforeUpload", "afterSelect", "onProgress", "onSuccess", "onError", "progressProps", "onSelect", "onChange", "onRemove", "onPreview", "sizeType", "sourceType", "children", "style"]);
  var locale = (0, _locale.useLocale)('Upload', others, _zhCn.default.Upload);

  var _useState = (0, _react.useState)('0px'),
      itemSize = _useState[0],
      setItemSize = _useState[1];

  var _useState2 = (0, _react.useState)(valueProp || defaultValue || []),
      value = _useState2[0],
      setValue = _useState2[1];

  var newestValue = (0, _react.useRef)(value);
  var timerQueue = (0, _react.useRef)([]);
  var uploadRef = (0, _react.useRef)();
  var uploadId = (0, _hooks.useGuid)('mtUpload');
  var uploadStyle = Object.assign({
    gridTemplateColumns: "repeat(" + cols + ", 1fr)"
  }, style);
  (0, _react.useEffect)(function () {
    if ((0, _utils.isValidArray)(valueProp)) {
      valueProp.forEach(function (file) {
        if (file.state && file.uid) {
          FILE_CACHAE[file.uid] = file.state;
        }
      });
    }

    setTimeout(function () {
      (function () {
        return (0, _tslib.__awaiter)(void 0, void 0, void 0, /*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
          var width, newCardSize;
          return _regeneratorRuntime().wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  _context.next = 2;
                  return (0, _utils.getWidth)(uploadId);

                case 2:
                  width = _context.sent;
                  newCardSize = "calc(((" + width + "px - var(--upload-card-spacing)*" + (cols - 1) + " )) / " + cols + ")";
                  setItemSize(newCardSize);

                case 5:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee);
        }));
      })();
    }, 200);
    return function () {
      timerQueue.current.forEach(function (x) {
        return clearInterval(x);
      });
      timerQueue.current = [];
    };
  }, []);
  var getFile = (0, _react.useCallback)(function (uid) {
    var item = newestValue.current.find(function (it) {
      return it.uid === uid;
    });

    if (item) {
      return {
        file: item,
        index: newestValue.current.indexOf(item)
      };
    }

    return undefined;
  }, []);
  var updateFile = (0, _react.useCallback)(function (uid, mod) {
    var gotFile = getFile(uid);

    if (!gotFile) {
      throw new Error("File id " + uid + " not found");
    }

    var file = gotFile.file,
        index = gotFile.index;

    if (mod.state) {
      FILE_CACHAE[uid] = mod.state;
    }

    var isSame = true;

    for (var k in mod) {
      if (mod.hasOwnProperty(k) && file[k] !== mod[k]) {
        isSame = false;
        break;
      }
    }

    if (isSame) {
      return file;
    }

    var newFile = Object.assign(Object.assign({}, file), mod);
    var newValue = [].concat(newestValue.current);
    newValue.splice(index, 1, newFile);
    newestValue.current = newValue;

    if (onChange) {
      onChange(newValue);
    }

    if (!('value' in props)) {
      setValue(newValue);
    }

    return newFile;
  }, [valueProp, onChange]);
  var uploadFiles = (0, _react.useCallback)(function (files) {
    var uploadFile = function uploadFile(it) {
      return (0, _tslib.__awaiter)(void 0, void 0, void 0, /*#__PURE__*/_regeneratorRuntime().mark(function _callee2() {
        var uid, options, res, curFile, timer, clearTimer, complete, newProcess;
        return _regeneratorRuntime().wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (!(it.state !== 'selected')) {
                  _context2.next = 2;
                  break;
                }

                return _context2.abrupt("return");

              case 2:
                uid = it.uid;
                updateFile(uid, {
                  state: 'pending'
                });
                options = {
                  action: action,
                  data: dataProp || {},
                  headers: headers || {}
                };

                if ((0, _utils.isFunction)(options.data)) {
                  options.data = options.data();
                }

                if (!beforeUpload) {
                  _context2.next = 29;
                  break;
                }

                _context2.prev = 7;
                res = beforeUpload(it, options);

                if (!(0, _utils.isBoolean)(res)) {
                  _context2.next = 15;
                  break;
                }

                if (!(res === false)) {
                  _context2.next = 13;
                  break;
                }

                updateFile(uid, {
                  state: 'uploading'
                });
                return _context2.abrupt("return");

              case 13:
                _context2.next = 22;
                break;

              case 15:
                if (!(0, _types.isPromise)(res)) {
                  _context2.next = 21;
                  break;
                }

                _context2.next = 18;
                return res;

              case 18:
                options = _context2.sent;
                _context2.next = 22;
                break;

              case 21:
                options = res;

              case 22:
                _context2.next = 29;
                break;

              case 24:
                _context2.prev = 24;
                _context2.t0 = _context2["catch"](7);
                updateFile(uid, {
                  state: 'error'
                });
                console.error(_context2.t0);
                return _context2.abrupt("return");

              case 29:
                curFile = updateFile(uid, {
                  state: 'uploading'
                });

                if (onProgress) {
                  onProgress();
                }

                if (options === null || options === void 0 ? void 0 : options.action) {
                  _context2.next = 36;
                  break;
                }

                console.warn('Upload request failed,  action prop is required');
                curFile = updateFile(uid, {
                  state: 'error'
                });

                if (_onError) {
                  _onError(curFile, newestValue.current);
                }

                return _context2.abrupt("return");

              case 36:
                timer = null;

                clearTimer = function clearTimer() {
                  if (timer && timerQueue.current.includes(timer)) {
                    clearInterval(timer);
                    timerQueue.current.splice(timerQueue.current.indexOf(timer), 1);
                  }

                  timer = null;
                };

                complete = function complete() {
                  return clearTimer();
                };

                newProcess = 0;
                _context2.next = 42;
                return new Promise(function (resolve, reject) {
                  var requestObj = {
                    onProgress: function onProgress(ev) {
                      if (curFile.state !== 'uploading') {
                        return;
                      }

                      if (!timer) {
                        timer = setInterval(function () {
                          if (curFile.state !== 'uploading') {
                            clearTimer();
                            return;
                          }

                          curFile = updateFile(uid, {
                            state: 'uploading',
                            percent: newProcess
                          });

                          if (newProcess === 100) {
                            clearTimer();
                          }
                        }, _utils2.UPLOAD_THROTTLE);
                        timerQueue.current.push(timer);
                      }

                      newProcess = ev.percent;
                    },
                    onError: function onError(err, body) {
                      console.error('Upload request failed', err);
                      curFile = updateFile(uid, {
                        state: 'error',
                        response: body
                      });

                      if (_onError) {
                        _onError(curFile, newestValue.current);
                      }

                      resolve('error');
                      complete();
                    },
                    onSuccess: function onSuccess(res) {
                      res.status = res.statusCode;

                      if ((0, _utils.isString)(res.data)) {
                        try {
                          res.data = JSON.parse(res.data);
                        } catch (e) {}
                      }

                      var result = formatter ? formatter(res, it) : res.data;

                      if (typeof result !== 'object') {
                        console.error("Expect a json object response, " + result + " got");
                        curFile = updateFile(uid, {
                          state: 'error'
                        });

                        if (_onError) {
                          _onError(curFile, newestValue.current);
                        }

                        resolve('error');
                        return;
                      }

                      var finalState = {
                        response: result,
                        percent: 100,
                        state: result.success ? 'done' : 'error'
                      };

                      if (result.success && result.url) {
                        finalState.url = result.url;
                        finalState.imgURL = result.imgURL || result.url;
                        finalState.downloadURL = result.downloadURL || result.url;
                      }

                      curFile = updateFile(uid, finalState);

                      if (finalState.state === 'done' && _onSuccess) {
                        _onSuccess(curFile, newestValue.current);
                      }

                      if (finalState.state === 'error' && _onError) {
                        _onError(curFile, newestValue.current);
                      }

                      clearTimer();
                      resolve('success');
                    },
                    fileType: fileType,
                    data: options.data,
                    filename: fileKeyName || nameProp,
                    file: it.file,
                    withCredentials: withCredentials,
                    action: options.action,
                    method: 'post',
                    timeout: 0,
                    headers: options.headers
                  };

                  if (request) {
                    request(requestObj);
                  } else {
                    (0, _request.default)(requestObj);
                  }
                });

              case 42:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, null, [[7, 24]]);
      }));
    };

    if (!(0, _utils.isValidArray)(files)) {
      return;
    }

    try {
      Promise.all(files.map(function (it) {
        return uploadFile(it);
      })).then(function () {
        var arr = updateFileStatus(newestValue.current);

        if (arr) {
          if ((0, _utils.isFunction)(onChange)) {
            onChange(arr);
          }

          if (!('value' in props)) {
            setValue(arr);
            newestValue.current = arr;
          }
        }
      });
    } catch (e) {
      console.warn('上传失败');
    }
  }, [action, dataProp, headers, beforeUpload, onProgress, formatter, _onSuccess, _onError, request]);
  var doUpload = (0, _react.useCallback)(function () {
    return uploadFiles(newestValue.current.filter(function (it) {
      return it.state === 'selected';
    }));
  }, []);
  var doReUpload = (0, _react.useCallback)(function (it) {
    updateFile(it.uid, {
      state: 'selected',
      percent: 0
    });
    doUpload();
  }, []);
  var handleAdd = (0, _react.useCallback)(function () {
    return (0, _tslib.__awaiter)(void 0, void 0, void 0, /*#__PURE__*/_regeneratorRuntime().mark(function _callee3() {
      var file, defaultSizeType, defaultSourceType, mediaType, fileItems, newValue;
      return _regeneratorRuntime().wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              if (!(limit <= newestValue.current.length || disabled)) {
                _context3.next = 3;
                break;
              }

              _message.default.notice((0, _utils.replaceMessage)(locale.limit, {
                limit: limit
              }));

              return _context3.abrupt("return");

            case 3:
              defaultSizeType = ['original', 'compressed'];
              defaultSourceType = ['album', 'camera'];
              _context3.prev = 5;

              if (!(0, _utils.isFunction)(filePicker)) {
                _context3.next = 12;
                break;
              }

              _context3.next = 9;
              return filePicker();

            case 9:
              file = _context3.sent;
              _context3.next = 23;
              break;

            case 12:
              if (!_utils.isStrictWechatMiniProgram) {
                _context3.next = 20;
                break;
              }

              mediaType = ['video', 'image'];

              if (accept === 'image/*') {
                mediaType = ['image'];
              } else if (accept === 'video/*') {
                mediaType = ['video'];
              }

              _context3.next = 17;
              return (0, _utils2.chooseImageForWechat)({
                count: Math.min(singleLimit, limit - newestValue.current.length),
                mediaType: mediaType,
                sizeType: sizeType || defaultSizeType,
                sourceType: sourceType || defaultSourceType
              });

            case 17:
              file = _context3.sent;
              _context3.next = 23;
              break;

            case 20:
              _context3.next = 22;
              return (0, _universalChooseImage.default)({
                count: Math.min(singleLimit, limit - newestValue.current.length),
                accept: accept,
                sizeType: sizeType || defaultSizeType,
                sourceType: sourceType || defaultSourceType
              });

            case 22:
              file = _context3.sent;

            case 23:
              _context3.next = 28;
              break;

            case 25:
              _context3.prev = 25;
              _context3.t0 = _context3["catch"](5);
              console.error(_context3.t0);

            case 28:
              if (!(!file || !Array.isArray(file === null || file === void 0 ? void 0 : file.data))) {
                _context3.next = 30;
                break;
              }

              return _context3.abrupt("return");

            case 30:
              fileItems = file.data.map(function (it, index) {
                return (0, _utils2.createFileItem)(it, file.files ? file.files[index] : null);
              });

              if (afterSelect) {
                fileItems = fileItems.filter(function (it) {
                  return afterSelect(it) !== false;
                });
              }

              if (useDataURL) {
                fileItems.forEach(function (it) {
                  if (it.dataURL && it.isImage) {
                    it.imgURL = it.dataURL;
                  }
                });
              }

              if (onSelect) {
                onSelect();
              }

              newValue = [].concat(newestValue.current, fileItems);

              if (onChange) {
                onChange(newValue);
              }

              if (!('value' in props)) {
                setValue(newValue);
                newestValue.current = newValue;
              }

              if (autoUpload) {
                doUpload();
              }

            case 38:
            case "end":
              return _context3.stop();
          }
        }
      }, _callee3, null, [[5, 25]]);
    }));
  }, [limit, disabled, locale, sizeType, sourceType, afterSelect, autoUpload, onSelect, onChange]);
  var handleRemove = (0, _react.useCallback)(function (file) {
    var doRemove = function doRemove() {
      var index = newestValue.current.indexOf(file);

      if (index === -1) {
        return;
      }

      var newValue = [].concat(newestValue.current);
      newValue.splice(index, 1);

      if (onChange) {
        onChange(newValue);
      }

      if (!('value' in props)) {
        setValue(newValue);
        newestValue.current = newValue;
      }
    };

    if (!onRemove) {
      doRemove();
      return;
    }

    var result = onRemove(file);

    if ((0, _utils.isBoolean)(result)) {
      if (result === true) {
        doRemove();
      }

      return;
    }

    if ((0, _types.isPromise)(result)) {
      result.then(function (should) {
        return should && doRemove();
      }).catch(function () {});
    }
  }, [onChange, onRemove]);
  var handleItemClick = (0, _react.useCallback)(function (file, e) {
    if (onPreview) {
      onPreview(file, Object.assign(Object.assign({}, e), {
        fileList: value
      }));
    }
  }, [onPreview]);
  (0, _react.useEffect)(function () {
    if ('value' in props) {
      newestValue.current = updateFileStatus(valueProp || defaultValue || []);
      setValue(newestValue.current);

      if (autoUpload) {
        doUpload();
      }
    }
  }, [valueProp]);
  (0, _react.useImperativeHandle)(ref, function () {
    return Object.assign(uploadRef.current, {
      add: handleAdd,
      upload: doUpload,
      getInstance: function getInstance() {
        return {
          startUpload: doUpload
        };
      }
    });
  });
  var clsPrefix = prefix + "upload";
  var hasAdd = showAddButton && limit > value.length && !disabled;
  return /*#__PURE__*/(0, _react.createElement)(_view.default, {
    id: uploadId,
    ref: uploadRef
  }, label || extraIcon ? /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: clsPrefix + "-header"
  }, /*#__PURE__*/(0, _react.createElement)(_meetReactComponentOne.Text, {
    className: clsPrefix + "-label " + clsPrefix + "-label--" + size
  }, label), extraIcon ? /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: clsPrefix + "-extra"
  }, /*#__PURE__*/(0, _react.createElement)(_meetReactComponentOne.Text, {
    className: clsPrefix + "-counter"
  }, value.length || 0, "/", limit), /*#__PURE__*/(0, _react.createElement)(_icon.default, {
    className: clsPrefix + "-extra-icon",
    size: size,
    onClick: handleAdd,
    name: extraIcon
  })) : null) : null, /*#__PURE__*/(0, _react.createElement)(_view.default, _extends({
    className: (0, _classnames.default)(clsPrefix, clsPrefix + "--" + listType, className),
    style: (0, _babelRuntimeJsxStyleTransform.rpx2vw4style)(uploadStyle)
  }, others), value.map(function (file) {
    return /*#__PURE__*/(0, _react.createElement)(_card.default, {
      key: file.uid,
      file: file,
      prefix: clsPrefix,
      progressProps: progressProps,
      onClick: handleItemClick,
      reUpload: locale.reupload,
      onReUpload: doReUpload,
      hasRemove: hasRemove && !disabled,
      onRemove: handleRemove,
      style: (0, _babelRuntimeJsxStyleTransform.rpx2vw4style)({
        width: "" + itemSize,
        height: "" + itemSize
      })
    });
  }), hasAdd && /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: (0, _classnames.default)(clsPrefix + "-item", clsPrefix + "-item--add", clsPrefix + "-item--" + listType),
    onClick: handleAdd,
    style: (0, _babelRuntimeJsxStyleTransform.rpx2vw4style)({
      width: "" + itemSize,
      height: "" + itemSize
    })
  }, children ? children : /*#__PURE__*/(0, _react.createElement)(_icon.default, {
    className: clsPrefix + "-item-icon",
    name: "add"
  }))));
};

Upload.displayName = 'Upload';

var _default = /*#__PURE__*/(0, _react.forwardRef)(Upload);

exports.default = _default;
