// @alifd/meet/lib/dialog/popup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var classnames_1 = tslib_1.__importDefault(require("classnames"));

var view_1 = tslib_1.__importDefault(require("../view"));

var modal_1 = tslib_1.__importDefault(require("../modal"));

var icon_1 = tslib_1.__importDefault(require("../icon"));

var utils_1 = require("../utils");

var Popup = function (props) {
  var _a = props.prefix,
      prefix = _a === void 0 ? 'mt-' : _a,
      _b = props.closeIconType,
      closeIconType = _b === void 0 ? 'error' : _b,
      _c = props.bgColor,
      bgColor = _c === void 0 ? 'transparent' : _c,
      visible = props.visible,
      children = props.children,
      _d = props.closeMode,
      closeMode = _d === void 0 ? ['mask', 'close'] : _d,
      className = props.className,
      style = props.style,
      onClose = props.onClose,
      others = tslib_1.__rest(props, ["prefix", "closeIconType", "bgColor", "visible", "children", "closeMode", "className", "style", "onClose"]);

  var clsPrefix = prefix + "dialog-popup";
  var containerStyle = style || {};

  if (bgColor) {
    containerStyle.backgroundColor = bgColor;
  }

  var handleClose = function (reason) {
    if (utils_1.isFunction(onClose) && Array.isArray(closeMode) && closeMode.indexOf(reason) > -1) {
      onClose(reason);
    }
  };

  return rax_1.createElement(modal_1.default, tslib_1.__assign({}, others, {
    visible: visible,
    containerClassName: clsPrefix,
    onClose: handleClose
  }), rax_1.createElement(view_1.default, {
    className: classnames_1.default(className, clsPrefix + "-container"),
    style: containerStyle
  }, children), rax_1.createElement(icon_1.default, {
    className: clsPrefix + "-close",
    type: closeIconType,
    size: "l",
    onClick: function onClick() {
      return handleClose('close');
    }
  }));
};

exports.default = Popup;
