// @alifd/meet/lib/select/section/select-search.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var search_1 = tslib_1.__importDefault(require("../../search"));

var utils_1 = require("../../utils");

var view_1 = tslib_1.__importDefault(require("../../view"));

var context_1 = tslib_1.__importDefault(require("../context"));

var SelectSearch = function (props, ref) {
  var _a = rax_1.useContext(context_1.default),
      prefix = _a.prefix,
      locale = _a.locale,
      setSearchValue = _a.setSearchValue,
      searchValue = _a.value,
      innerValue = _a.innerValue,
      contextSearch = _a.onSearch;

  var onSearch = props.onSearch;

  var _b = rax_1.useState(''),
      value = _b[0],
      _setValue = _b[1];

  rax_1.useImperativeHandle(ref, function () {
    return {
      setValue: function setValue(v) {
        _setValue(v);

        setSearchValue(v);
      }
    };
  });
  return rax_1.createElement(view_1.default, {
    className: prefix + "select-search-wrapper",
    style: {
      width: '100%',
      height: '100rpx'
    }
  }, rax_1.createElement(search_1.default, {
    placeholder: locale.searchTipText,
    value: value,
    hasClear: true,
    onChange: function handleSearch(v) {
      _setValue(v);

      setSearchValue(v);

      if (utils_1.isFunction(onSearch)) {
        onSearch(v);
      }

      if (utils_1.isFunction(contextSearch)) {
        contextSearch(v, searchValue, innerValue);
      }
    }
  }));
};

exports.default = rax_1.forwardRef(SelectSearch);
