// whale-component-docgen/src/render/renderPackageJSON/renderComponentPackageJSON.ts
import { code, heading, paragraph, text, link } from 'mdast-builder';
import { MaterialPackageJSON } from '../../types';
import * as qs from 'qs';
import { getGitRemoteUrl } from '../../utils/remoteUrl';
import {
  gitlabBugIssueBadge,
  gitlabEnhancementIssueBadge,
  tnpmDownloadsBadge,
  tnpmVersionBadge,
} from './renderBadges';

export const renderComponentPackageJSON = (
  packageJSON: MaterialPackageJSON,
  source: string
) =>
  [
    heading(1, text(packageJSON.name)),
    paragraph(renderBadges(packageJSON, source)),
    paragraph(text(packageJSON.description)),
    heading(2, text('安装方法')),
    code('sh', `$ tnpm install ${packageJSON.name} --save`),
    packageJSON.componentConfig &&
      paragraph(
        link(
          'https://yuque.antfin-inc.com/bp-xt-dt/battle/components_used_in_iceworks',
          '如何iceworks中使用组件',
          text('如何iceworks中使用组件')
        )
      ),
    heading(2, text('API')),
  ].filter((vo) => vo);

const renderBadges = (packageJSON: MaterialPackageJSON, source: string) => {
  const remoteUrl = getGitRemoteUrl(source);
  const materialConfig = packageJSON.componentConfig;
  return [
    tnpmVersionBadge(packageJSON.name, packageJSON.version),
    text('\n'),
    tnpmDownloadsBadge(packageJSON.name),
    text('\n'),
    remoteUrl &&
      gitlabBugIssueBadge(
        `${remoteUrl}/issues/new?${qs.stringify({
          issue: {
            title: `[Bug]${materialConfig?.name || packageJSON.name}: `,
            description: [
              '### 环境信息',
              '',
              `- 组件名称： ${materialConfig?.title}`,
              `- 组件版本： ${packageJSON.version}`,
              '',
              '<!-- 在此描述Bug -->',
              '',
              '### 当前行为',
              '',
              '### 期望行为',
              '',
              '### 复现方式',
              '',
            ].join('\n'),
            issuable_template: 'bug',
          },
        })}`,
        '提Bug'
      ),
    remoteUrl && text('\n'),
    remoteUrl &&
      gitlabEnhancementIssueBadge(
        `${remoteUrl}/issues/new?${qs.stringify({
          issue: {
            title: `[需求]${materialConfig?.name || packageJSON.name}: `,
            description: [
              '### 环境信息',
              '',
              `- 组件名称: ${materialConfig?.name} ${materialConfig?.title}`,
              `- 组件版本: ${packageJSON.version}`,
              '',
              '<!-- 在此描述需求 -->',
              '',
              '### 期望行为',
              '',
              '### UI截图(如果有设计稿请附上)',
              '',
            ].join('\n'),
            issuable_template: 'enhancement',
          },
        })}`,
        '提需求'
      ),

    // https://teambition.alibaba-inc.com/project/5edf236764aa23bc59c3c4c4/tasks/scrum/5edf23683af5f1af658f450c
  ].filter((vo) => vo);
};
