// @alilc/lowcode-rax-renderer/es/index.js
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
import { Component, PureComponent, createElement, createContext, forwardRef } from 'rax';
import findDOMNode from 'rax-find-dom-node';
import { adapter, addonRendererFactory, tempRendererFactory, rendererFactory } from '@alilc/lowcode-renderer-core';
import pageRendererFactory from './renderer/page';
import componentRendererFactory from './renderer/component';
import blockRendererFactory from './renderer/block';
import CompFactory from './hoc/compFactory';
adapter.setRuntime({
  Component: Component,
  PureComponent: PureComponent,
  createContext: createContext,
  createElement: createElement,
  forwardRef: forwardRef,
  findDOMNode: findDOMNode
});
adapter.setRenderers({
  PageRenderer: pageRendererFactory(),
  ComponentRenderer: componentRendererFactory(),
  BlockRenderer: blockRendererFactory(),
  AddonRenderer: addonRendererFactory(),
  TempRenderer: tempRendererFactory()
});
function factory() {
  var Renderer = rendererFactory();
  return /*#__PURE__*/function (_Renderer) {
    _inheritsLoose(_class, _Renderer);
    function _class(props, context) {
      return _Renderer.call(this, props, context) || this;
    }
    var _proto = _class.prototype;
    _proto.isValidComponent = function isValidComponent(obj) {
      var _obj$prototype;
      return (obj == null ? void 0 : (_obj$prototype = obj.prototype) == null ? void 0 : _obj$prototype.setState) || (obj == null ? void 0 : obj.prototype) instanceof Component;
    };
    return _class;
  }(Renderer);
}
var RaxRenderer = factory();
var Engine = RaxRenderer;
export { Engine, CompFactory };
export default RaxRenderer;
