// @alilc/lowcode-material-parser/lib/parse/dynamic/requireInSandbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = require("fs-extra");
const vm2_1 = require("vm2");
// import PropTypes from 'prop-types';
const cssPattern = /\.(css|scss|sass|less)$/;
function requireInSandbox(filePath, PropTypes) {
    const vm = new vm2_1.NodeVM({
        sandbox: {},
        sourceExtensions: ['js', 'css', 'scss', 'sass', 'less'],
        compiler: (code, filename) => {
            if (filename.match(cssPattern)) {
                return `
              const handler = {
                get() {
                  return new Proxy({}, handler);
                },
              };
              const proxiedObject = new Proxy({}, handler);
              module.exports = proxiedObject;
            `;
            }
            else {
                return code;
            }
        },
        require: {
            external: true,
            context: 'sandbox',
            mock: {
                'prop-types': PropTypes,
            },
        },
    });
    const fileContent = fs_extra_1.readFileSync(filePath, { encoding: 'utf8' });
    return vm.run(fileContent, filePath);
}
exports.default = requireInSandbox;
