// rax-miniapp-babel-plugins/es/plugins/babel-plugin-handle-plugin-component.js
"use strict";

var _require = require('../utils/handleComponentAST'),
    collectComponentAttr = _require.collectComponentAttr,
    collectUsings = _require.collectUsings;

var MINIAPP_PLUGIN_COMPONENTS_REG = /^plugin\:\/\//;

module.exports = function visitor(_ref, _ref2) {
  var t = _ref.types;
  var usingPlugins = _ref2.usingPlugins;
  // Collect imported dependencies
  var pluginComponents = {};
  var scanedFileMap = {};
  return {
    visitor: {
      Program: {
        exit: function exit(path, _ref3) {
          var filename = _ref3.filename;
          scanedFileMap[filename] = false;
          pluginComponents = {};
        }
      },
      ImportDeclaration: {
        enter: function enter(path, _ref4) {
          var filename = _ref4.filename;
          var source = path.node.source;

          if (t.isStringLiteral(source) && MINIAPP_PLUGIN_COMPONENTS_REG.test(source.value)) {
            if (!scanedFileMap[filename]) {
              scanedFileMap[filename] = true;
              path.parentPath.traverse({
                JSXOpeningElement: collectComponentAttr(pluginComponents, t)
              });
            }

            collectUsings(path, pluginComponents, usingPlugins, source.value, t);
          }
        }
      }
    }
  };
};
