// rax-video/es/web/index.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import { createElement, useRef, useEffect, forwardRef, useImperativeHandle } from 'rax';
import cx from 'classnames/dedupe';
import omit from 'omit.js';
import wrapDefaultProperties from '../utils/wrapDefaultProperties';
var noop = function () {};
var Video = forwardRef(function (props, ref) {
  var className = props.className,
    style = props.style,
    controls = props.controls,
    playControl = props.playControl,
    autoPlay = props.autoPlay,
    _props$onPlayError = props.onPlayError,
    onPlayError = _props$onPlayError === void 0 ? noop : _props$onPlayError;
  var refEl = useRef(null);
  useImperativeHandle(ref, function () {
    return refEl.current;
  });
  var common = omit(props, ['className', 'controls', 'style', 'playControl']);
  // Default controls is true
  if (controls == undefined || controls === true) {
    common.controls = true;
  } else {
    common.controls = false;
  }
  common.autoPlay = playControl === 'play' || autoPlay;
  if (common.autoPlay === false) {
    delete common.autoPlay; // In W3C standard, if the attribute is set, it will be treated as true regardless of its value
  }

  useEffect(function () {
    var node = refEl.current;
    if (playControl !== undefined) {
      if (playControl === 'play') {
        var playPromise = node.play(); // Should return a Promise https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play
        if (playPromise !== undefined) {
          playPromise.catch(function (error) {
            onPlayError(error);
          });
        }
      } else {
        node.pause();
      }
    }
  }, [playControl]);
  return createElement("video", _extends({}, common, {
    ref: refEl,
    className: cx('rax-video', className),
    style: style,
    "webkit-playsinline": true,
    playsinline: true,
    src: props.src
  }));
});
export default wrapDefaultProperties(Video);
