// @alilc/lowcode-react-renderer/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports["default"] = void 0;
var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));
var _react = _interopRequireWildcard(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _lowcodeRendererCore = require("@alilc/lowcode-renderer-core");
var _configProvider = _interopRequireDefault(require("@alifd/next/lib/config-provider"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
window.React = _react["default"];
window.ReactDom = _reactDom["default"];
_lowcodeRendererCore.adapter.setRuntime({
  Component: _react.Component,
  PureComponent: _react.PureComponent,
  createContext: _react.createContext,
  createElement: _react.createElement,
  forwardRef: _react.forwardRef,
  findDOMNode: _reactDom["default"].findDOMNode
});
_lowcodeRendererCore.adapter.setRenderers({
  PageRenderer: (0, _lowcodeRendererCore.pageRendererFactory)(),
  ComponentRenderer: (0, _lowcodeRendererCore.componentRendererFactory)(),
  BlockRenderer: (0, _lowcodeRendererCore.blockRendererFactory)(),
  AddonRenderer: (0, _lowcodeRendererCore.addonRendererFactory)(),
  TempRenderer: (0, _lowcodeRendererCore.tempRendererFactory)(),
  DivRenderer: (0, _lowcodeRendererCore.blockRendererFactory)()
});
_lowcodeRendererCore.adapter.setConfigProvider(_configProvider["default"]);
function factory() {
  var Renderer = (0, _lowcodeRendererCore.rendererFactory)();
  return /*#__PURE__*/function (_Renderer) {
    (0, _inheritsLoose2["default"])(ReactRenderer, _Renderer);
    function ReactRenderer(props, context) {
      var _this;
      _this = _Renderer.call(this, props, context) || this;
      _this.props = void 0;
      _this.context = void 0;
      _this.setState = void 0;
      _this.forceUpdate = void 0;
      _this.refs = void 0;
      return _this;
    }
    var _proto = ReactRenderer.prototype;
    _proto.isValidComponent = function isValidComponent(obj) {
      var _obj$prototype;
      return (obj === null || obj === void 0 ? void 0 : (_obj$prototype = obj.prototype) === null || _obj$prototype === void 0 ? void 0 : _obj$prototype.isReactComponent) || (obj === null || obj === void 0 ? void 0 : obj.prototype) instanceof _react.Component;
    };
    return ReactRenderer;
  }(Renderer);
}
var _default = factory();
exports["default"] = _default;
