// @alifd/meet/lib/progress/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.progressCircleFontSize = exports.progressCircleWidthSize = exports.progressiveColor = exports.progressSize = exports.progressBackgroundColor = exports.statusColor = exports.getValue = exports.addRpx = void 0;

var utils_1 = require("../utils");

exports.addRpx = function (value) {
  if (!value) {
    return value;
  }

  return utils_1.isString(value) ? value : value + "rpx";
};

exports.getValue = function (cssProp) {
  if (!cssProp) {
    return 0;
  }

  if (utils_1.isNumber(cssProp)) {
    return cssProp;
  }

  var index = cssProp.indexOf('rpx');

  if (index === -1) {
    return Number(cssProp);
  }

  return Number(cssProp.substring(0, index));
};

exports.statusColor = function (status) {
  return {
    normal: '#209BFA',
    error: '#EB2F2F',
    success: '#05D161'
  }[status];
};

exports.progressBackgroundColor = function (color) {
  return color || '#f2f2f2';
};

exports.progressSize = function (size) {
  return {
    small: 4,
    medium: 6,
    large: 8
  }[size] || size;
};

exports.progressiveColor = function (percent) {
  if (percent < 33) {
    return '#EB2F2F';
  }

  if (percent > 66) {
    return '#05D161';
  }

  return '#D97B00';
};

exports.progressCircleWidthSize = function (size) {
  return {
    small: 22,
    medium: 28,
    large: 34
  }[size] || size;
};

exports.progressCircleFontSize = function (size) {
  return {
    small: 14,
    medium: 16,
    large: 18
  }[size] || size;
};
