// @alifd/meet/lib/checkbox-field/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var checkbox_1 = tslib_1.__importDefault(require("../checkbox"));

var field_1 = require("../field");

var CheckboxField = function (_super) {
  tslib_1.__extends(CheckboxField, _super);

  function CheckboxField() {
    return _super !== null && _super.apply(this, arguments) || this;
  }

  CheckboxField.prototype.renderPreview = function (value) {
    var _a = this.props,
        _b = _a.direction,
        direction = _b === void 0 ? 'hoz' : _b,
        others = tslib_1.__rest(_a, ["direction"]);

    return rax_1.createElement(checkbox_1.default.Group, tslib_1.__assign({
      value: value,
      style: {
        maxWidth: '100%'
      },
      direction: direction,
      disabled: true
    }, others));
  };

  CheckboxField.prototype.renderControl = function () {
    var _a = this.props,
        _b = _a.direction,
        direction = _b === void 0 ? 'hoz' : _b,
        others = tslib_1.__rest(_a, ["direction"]);

    return rax_1.createElement(checkbox_1.default.Group, tslib_1.__assign({
      style: {
        maxWidth: '100%'
      },
      direction: direction
    }, others));
  };

  CheckboxField.displayName = 'CheckboxField';
  return CheckboxField;
}(field_1.FieldElement);

exports.default = CheckboxField;
