// @alifd/meet/lib/date-picker/drawer-picker/minute-picker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var minute_picker_1 = tslib_1.__importDefault(require("../embed-picker/minute-picker"));

var base_picker_1 = tslib_1.__importDefault(require("./base-picker"));

var MinutePicker = function (_super) {
  tslib_1.__extends(MinutePicker, _super);

  function MinutePicker() {
    return _super !== null && _super.apply(this, arguments) || this;
  }

  MinutePicker.prototype.renderPicker = function () {
    var validRange = this.props.validRange;
    return rax_1.createElement(minute_picker_1.default, {
      value: this.state.pickerValue,
      disabledDate: this.props.disabledDate,
      validRange: validRange,
      onChange: this.handleChange
    });
  };

  Object.defineProperty(MinutePicker.prototype, "format", {
    get: function get() {
      return this.props.format;
    },
    enumerable: false,
    configurable: true
  });

  MinutePicker.prototype.getDefaultFormat = function () {
    return '';
  };

  MinutePicker.defaultProps = {
    format: 'YYYY-MM-DD HH:mm'
  };
  return MinutePicker;
}(base_picker_1.default);

exports.default = MinutePicker;
