// @alifd/meet/lib/date-picker/drawer-picker/month-range-picker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var month_picker_1 = tslib_1.__importDefault(require("../embed-picker/month-picker"));

var base_range_picker_1 = tslib_1.__importDefault(require("./base-range-picker"));

var locale_1 = require("../../locale");

var MonthRangePicker = function (_super) {
  tslib_1.__extends(MonthRangePicker, _super);

  function MonthRangePicker() {
    return _super !== null && _super.apply(this, arguments) || this;
  }

  Object.defineProperty(MonthRangePicker.prototype, "rangeType", {
    get: function get() {
      return 'month';
    },
    enumerable: false,
    configurable: true
  });

  MonthRangePicker.prototype.renderFirstChild = function (value, validRange, onChange) {
    return this.renderBase(value, validRange, onChange);
  };

  MonthRangePicker.prototype.renderSecondChild = function (value, validRange, onChange) {
    return this.renderBase(value, validRange, onChange);
  };

  MonthRangePicker.prototype.renderBase = function (value, validRange, onChange) {
    return rax_1.createElement(month_picker_1.default, {
      value: value,
      validRange: validRange,
      onChange: onChange
    });
  };

  Object.defineProperty(MonthRangePicker.prototype, "format", {
    get: function get() {
      return this.props.format || 'YYYY-MM';
    },
    enumerable: false,
    configurable: true
  });
  return MonthRangePicker;
}(base_range_picker_1.default);

exports.default = locale_1.withLocale(MonthRangePicker, {
  name: 'MonthRangePicker',
  defaultLocale: {
    ok: '',
    cancel: '',
    prev: '上一步',
    next: '下一步',
    selectStartTime: '选择开始日期',
    selectEndTime: '选择结束日期'
  }
});
