// @alifd/meet/lib/list/item/media.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListItemAction = exports.ListItemMedia = void 0;

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var classnames_1 = tslib_1.__importDefault(require("classnames"));

var view_1 = tslib_1.__importDefault(require("../../view"));

var icon_1 = tslib_1.__importDefault(require("../../icon"));

var Media = function (props, ref) {
  var action = props.action,
      _a = props.prefix,
      prefix = _a === void 0 ? 'mt-' : _a,
      icon = props.icon,
      children = props.children,
      className = props.className,
      others = tslib_1.__rest(props, ["action", "prefix", "icon", "children", "className"]);

  return rax_1.createElement(view_1.default, tslib_1.__assign({
    ref: ref
  }, others, {
    className: classnames_1.default(prefix + "list-item-media", action ? prefix + "list-item-media--last" : null, className)
  }), icon ? rax_1.createElement(icon_1.default, {
    className: prefix + "list-item-icon",
    name: icon
  }) : null, rax_1.createElement(rax_1.Fragment, null, children));
};

exports.ListItemMedia = rax_1.forwardRef(Media);

var Action = function (props, ref) {
  return rax_1.createElement(exports.ListItemMedia, tslib_1.__assign({}, props, {
    action: true,
    ref: ref
  }));
};

exports.ListItemAction = rax_1.forwardRef(Action);
exports.default = exports.ListItemMedia;
