// @alifd/meet/lib/menu/sub-menu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var classnames_1 = tslib_1.__importDefault(require("classnames"));

var list_1 = tslib_1.__importDefault(require("../list"));

var view_1 = tslib_1.__importDefault(require("../view"));

var menu_context_1 = tslib_1.__importDefault(require("./menu-context"));

var constant_1 = require("./constant");

var SubMenu = function (props, ref) {
  var _a, _b;

  var context = rax_1.useContext(menu_context_1.default);

  var className = props.className,
      containerClassName = props.containerClassName,
      containerStyle = props.containerStyle,
      children = props.children,
      label = props.label,
      loading = props.loading,
      keyPath = props.keyPath,
      _c = props.depth,
      depth = _c === void 0 ? 1 : _c,
      others = tslib_1.__rest(props, ["className", "containerClassName", "containerStyle", "children", "label", "loading", "keyPath", "depth"]);

  var prefix = context.prefix,
      _d = context.inlineIndent,
      inlineIndent = _d === void 0 ? constant_1.DEFAULT_INLINE_INDENT : _d,
      openKeys = context.openKeys,
      onOpen = context.onOpen;
  var open = keyPath && openKeys.includes(keyPath);
  return rax_1.createElement(view_1.default, tslib_1.__assign({}, others, {
    ref: ref,
    className: containerClassName,
    style: containerStyle
  }), rax_1.createElement(list_1.default.Item, {
    className: classnames_1.default(prefix + "menu-sub-menu", className, (_a = {}, _a[prefix + "menu-sub-menu--loading"] = loading, _a)),
    title: label,
    arrow: loading ? 'loading' : open ? 'up' : 'down',
    onClick: function handleOpen() {
      if (!loading) {
        onOpen({
          key: keyPath,
          open: !open
        });
      }
    }
  }), rax_1.createElement(view_1.default, {
    className: classnames_1.default(prefix + "menu-sub-menu-list", (_b = {}, _b[prefix + "menu-sub-menu-list--open"] = open, _b)),
    style: {
      textIndent: inlineIndent * 2 * depth + "rpx"
    }
  }, children));
};

exports.default = rax_1.forwardRef(SubMenu);
