// rax-miniapp-runtime-webpack-plugin/es/generators/config.js
"use strict";

var addFileToCompilation = require('../utils/addFileToCompilation');

module.exports = function (compilation, _ref) {
  var usingComponents = _ref.usingComponents,
      usingPlugins = _ref.usingPlugins,
      pages = _ref.pages,
      target = _ref.target,
      command = _ref.command;
  var config = {
    usingComponents: {},
    usingPlugins: {},
    pages: pages
  };

  if (process.env.DEBUG === 'true') {
    config.debug = true;
  }

  Object.keys(usingComponents).forEach(function (name) {
    config.usingComponents[name] = {
      props: usingComponents[name].props,
      events: usingComponents[name].events
    };
  });
  Object.keys(usingPlugins).forEach(function (name) {
    config.usingPlugins[name] = {
      props: usingPlugins[name].props,
      events: usingPlugins[name].events
    };
  });
  addFileToCompilation(compilation, {
    filename: 'config.js',
    content: "module.exports = " + JSON.stringify(config),
    command: command,
    target: target
  });
};
