// rax-miniapp-runtime-webpack-plugin/es/generators/index.js
"use strict";

var _require = require('./app'),
    generateAppJS = _require.generateAppJS,
    generateAppCSS = _require.generateAppCSS;

var generateConfig = require('./config');

var _require2 = require('./page'),
    generatePageCSS = _require2.generatePageCSS,
    generatePageJS = _require2.generatePageJS,
    generatePageJSON = _require2.generatePageJSON,
    generatePageXML = _require2.generatePageXML;

var _require3 = require('./root'),
    generateRootTmpl = _require3.generateRootTmpl;

var _require4 = require('./element'),
    generateElementJS = _require4.generateElementJS,
    generateElementJSON = _require4.generateElementJSON,
    generateElementTemplate = _require4.generateElementTemplate;

var generateRender = require('./render');

var generatePkg = require('./pkg');

module.exports = {
  generateAppCSS: generateAppCSS,
  generateAppJS: generateAppJS,
  generateConfig: generateConfig,
  generatePageCSS: generatePageCSS,
  generatePageJS: generatePageJS,
  generatePageJSON: generatePageJSON,
  generatePageXML: generatePageXML,
  generateRootTmpl: generateRootTmpl,
  generateElementJS: generateElementJS,
  generateElementJSON: generateElementJSON,
  generateElementTemplate: generateElementTemplate,
  generateRender: generateRender,
  generatePkg: generatePkg
};
