// rax-miniapp-runtime-webpack-plugin/es/generators/root.js
"use strict";

var adapter = require('../adapter');

var ejs = require('ejs');

var addFileToCompilation = require('../utils/addFileToCompilation');

var getTemplate = require('../utils/getTemplate');

function generateRootTmpl(compilation, _ref) {
  var usingPlugins = _ref.usingPlugins,
      usingComponents = _ref.usingComponents,
      target = _ref.target,
      command = _ref.command,
      rootDir = _ref.rootDir;
  var template = ejs.render(getTemplate(rootDir, 'root.xml', target));
  var pluginTmpl = ejs.render(getTemplate(rootDir, 'plugin.xml', target), {
    usingPlugins: usingPlugins
  });
  var componentTmpl = ejs.render(getTemplate(rootDir, 'custom-component.xml', target), {
    usingComponents: usingComponents
  });
  addFileToCompilation(compilation, {
    filename: "root." + adapter[target].xml,
    content: template + pluginTmpl + componentTmpl,
    target: target,
    command: command
  });
  addFileToCompilation(compilation, {
    filename: "tool." + adapter[target].script,
    content: ejs.render(getTemplate(rootDir, "tool." + adapter[target].script, target)),
    target: target,
    command: command
  });
}

module.exports = {
  generateRootTmpl: generateRootTmpl
};
