// rax-canvas/es/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import { createElement, Component, createRef } from 'rax';
import { isWeex } from 'universal-env';
import { enable, WeexBridge, Image as GImage } from '@gcanvas/core';
import findDOMNode from 'rax-find-dom-node';

var Canvas = /*#__PURE__*/function (_Component) {
  _inheritsLoose(Canvas, _Component);

  function Canvas(props) {
    var _this = _Component.call(this, props) || this;

    _this.getContext = function (type) {
      if (type === void 0) {
        type = '2d';
      }

      var canvas = findDOMNode(_this.canvas.current);

      if (isWeex) {
        var gcanvas = enable(canvas, {
          bridge: WeexBridge,
          debug: false,
          disableAutoSwap: false,
          disableComboCommands: false
        });
        return gcanvas.getContext(type);
      } else {
        return canvas.getContext(type);
      }
    };

    _this.canvas = createRef();
    return _this;
  }

  var _proto = Canvas.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        _this$props$style = _this$props.style,
        style = _this$props$style === void 0 ? {} : _this$props$style,
        width = _this$props.width,
        height = _this$props.height;
    var customStyle = {
      width: width !== undefined ? width : style.width,
      height: height !== undefined ? height : style.height
    };

    if (isWeex) {
      return createElement("gcanvas", _extends({}, this.props, {
        style: _extends({}, style, customStyle),
        ref: this.canvas
      }));
    } else {
      return createElement("canvas", _extends({}, this.props, {
        width: parseInt(customStyle.width),
        height: parseInt(customStyle.height),
        ref: this.canvas
      }));
    }
  };

  return Canvas;
}(Component);

Canvas.createImage = function () {
  if (isWeex) {
    return new GImage();
  } else {
    return new Image();
  }
};

export default Canvas;
