// @alifd/meet/lib/action-sheet/open.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var render_1 = require("../utils/render");

var utils_1 = require("../utils");

var action_sheet_1 = tslib_1.__importDefault(require("./action-sheet"));

var ActionSheetWrapper = function (_super) {
  tslib_1.__extends(ActionSheetWrapper, _super);

  function ActionSheetWrapper(props) {
    var _this = _super.call(this, props) || this;

    _this.handleClick = function (option, index, e) {
      if (_this.props.onClick) {
        _this.props.onClick(option, index, e);
      }
    };

    _this.handleClose = function (reason) {
      _this.setState({
        visible: false
      });

      if (_this.props.onClose) {
        _this.props.onClose(reason);
      }
    };

    _this.state = {
      visible: true
    };
    return _this;
  }

  ActionSheetWrapper.prototype.render = function () {
    var _a = this.props,
        message = _a.message,
        options = _a.options,
        others = tslib_1.__rest(_a, ["message", "options"]);

    var visible = this.state.visible;
    return rax_1.createElement(action_sheet_1.default, tslib_1.__assign({
      visible: visible
    }, others, {
      disablePortal: true,
      message: message,
      options: options,
      onClose: this.handleClose,
      onClick: this.handleClick
    }));
  };

  ActionSheetWrapper.defaultProps = {
    message: '',
    options: [],
    onClick: function onClick() {},
    onClose: function onClose() {}
  };
  return ActionSheetWrapper;
}(rax_1.Component);

function default_1(options) {
  var onClose = options.onClose,
      others = tslib_1.__rest(options, ["onClose"]);

  var hide = null;
  hide = render_1.renderToRoot(rax_1.createElement(ActionSheetWrapper, tslib_1.__assign({}, others, {
    onClose: function handleClose(v) {
      if (utils_1.isFunction(onClose)) {
        onClose(v);
      }

      setTimeout(function () {
        hide();
      }, 300);
    }
  })));
  return {
    close: function close() {
      return hide();
    }
  };
}

exports.default = default_1;
