// @uni/system-info/es/web/getInfoSync.js
import { formatSystemInfo } from "../common";
function getPlatform() {
  var _navigator = navigator,
    platform = _navigator.platform;
  if (platform) {
    var _navigator2, _navigator2$userAgent, _navigator3, _navigator3$userAgent;
    if (/(iphone|ipod|ipad)/.test((_navigator2 = navigator) == null ? void 0 : (_navigator2$userAgent = _navigator2.userAgent) == null ? void 0 : _navigator2$userAgent.toLowerCase())) return 'iOS';
    if (/android/.test((_navigator3 = navigator) == null ? void 0 : (_navigator3$userAgent = _navigator3.userAgent) == null ? void 0 : _navigator3$userAgent.toLowerCase())) return 'Android';
    return platform;
  }
  return '';
}
var getInfoSync = formatSystemInfo(function () {
  return {
    pixelRatio: window.devicePixelRatio,
    screenWidth: window.document.documentElement.clientWidth,
    screenHeight: window.document.documentElement.clientHeight,
    windowWidth: window.innerWidth,
    windowHeight: window.innerHeight,
    language: navigator.language,
    version: navigator.appVersion,
    platform: getPlatform()
  };
});
export default getInfoSync;
