// @uni/system-info/es/web/getInfo.js
import { formatSystemInfo, formatSystemInfoAsync } from "../common";
import { CONTAINER_NAME } from "../_utils/constant";
function getPlatform() {
  var _navigator = navigator,
    platform = _navigator.platform;
  if (platform) {
    var _navigator2, _navigator2$userAgent, _navigator3, _navigator3$userAgent;
    if (/(iphone|ipod|ipad)/.test((_navigator2 = navigator) == null ? void 0 : (_navigator2$userAgent = _navigator2.userAgent) == null ? void 0 : _navigator2$userAgent.toLowerCase())) return 'iOS';
    if (/android/.test((_navigator3 = navigator) == null ? void 0 : (_navigator3$userAgent = _navigator3.userAgent) == null ? void 0 : _navigator3$userAgent.toLowerCase())) return 'Android';
    return platform;
  }
  return '';
}
var getInfoSync = formatSystemInfo(function () {
  return {
    pixelRatio: window.devicePixelRatio,
    screenWidth: window.document.documentElement.clientWidth,
    screenHeight: window.document.documentElement.clientHeight,
    windowWidth: window.innerWidth,
    windowHeight: window.innerHeight,
    language: navigator.language,
    version: navigator.appVersion,
    platform: getPlatform() || ''
  };
});
var getInfo = formatSystemInfoAsync(function (options) {
  var success = options.success,
    fail = options.fail,
    complete = options.complete;
  try {
    var res = getInfoSync();
    success && success(res);
    complete && complete(res);
  } catch (e) {
    fail && fail(e);
    complete && complete(e);
  }
}, CONTAINER_NAME.WEB);
export default getInfo;
