// @alilc/lowcode-types/lib/assets.js
"use strict";

exports.__esModule = true;
exports.AssetType = exports.AssetLevels = exports.AssetLevel = void 0;
var AssetLevel;
exports.AssetLevel = AssetLevel;
(function (AssetLevel) {
  AssetLevel[AssetLevel["Environment"] = 1] = "Environment";
  AssetLevel[AssetLevel["Library"] = 2] = "Library";
  AssetLevel[AssetLevel["Theme"] = 3] = "Theme";
  AssetLevel[AssetLevel["Runtime"] = 4] = "Runtime";
  AssetLevel[AssetLevel["Components"] = 5] = "Components";
  AssetLevel[AssetLevel["App"] = 6] = "App";
})(AssetLevel || (exports.AssetLevel = AssetLevel = {}));
var AssetLevels = [AssetLevel.Environment, AssetLevel.Library, AssetLevel.Theme, AssetLevel.Runtime, AssetLevel.Components, AssetLevel.App];
exports.AssetLevels = AssetLevels;
var AssetType;
exports.AssetType = AssetType;
(function (AssetType) {
  AssetType["JSUrl"] = "jsUrl";
  AssetType["CSSUrl"] = "cssUrl";
  AssetType["CSSText"] = "cssText";
  AssetType["JSText"] = "jsText";
  AssetType["Bundle"] = "bundle";
})(AssetType || (exports.AssetType = AssetType = {}));
