// @alifd/meet/lib/radio-field/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var tslib_1 = require("tslib");

var rax_1 = require("rax");

var field_1 = require("../field");

var radio_1 = tslib_1.__importDefault(require("../radio"));

var RadioField = function (_super) {
  tslib_1.__extends(RadioField, _super);

  function RadioField() {
    return _super !== null && _super.apply(this, arguments) || this;
  }

  RadioField.prototype.renderPreview = function (value) {
    var _a = this.props,
        _b = _a.direction,
        direction = _b === void 0 ? 'hoz' : _b,
        others = tslib_1.__rest(_a, ["direction"]);

    return rax_1.createElement(radio_1.default.Group, tslib_1.__assign({
      value: value,
      style: {
        maxWidth: '100%'
      },
      direction: direction,
      disabled: true
    }, others));
  };

  RadioField.prototype.renderControl = function () {
    var _a = this.props,
        _b = _a.direction,
        direction = _b === void 0 ? 'hoz' : _b,
        others = tslib_1.__rest(_a, ["direction"]);

    return rax_1.createElement(radio_1.default.Group, tslib_1.__assign({
      style: {
        maxWidth: '100%'
      },
      direction: direction
    }, others));
  };

  RadioField.displayName = 'RadioField';
  return RadioField;
}(field_1.FieldElement);

exports.default = RadioField;
