// build-plugin-component-multiple/src/utils/css/setCSSRule.js
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const getPostcssPlugins = require('./getPostcssPlugins');

module.exports = (config, useStyleLoader = true) => {
  if (useStyleLoader) {
    config.module
      .rule('css')
      .test(/\.css?$/)
      .use('style')
      .loader(require.resolve('style-loader'))
      .end()
      .use('css')
      .loader(require.resolve('css-loader'))
      .end()
      .use('postcss')
      .loader(require.resolve('postcss-loader'))
      .options({
        ident: 'postcss',
        // TODO: postcss 待升级到 8.x
        plugins: () => getPostcssPlugins(),
      })
      .end();

    config.module
      .rule('less')
      .test(/\.less?$/)
      .use('style')
      .loader(require.resolve('style-loader'))
      .end()
      .use('css')
      .loader(require.resolve('css-loader'))
      .end()
      .use('postcss')
      .loader(require.resolve('postcss-loader'))
      .options({
        ident: 'postcss',
        plugins: () => getPostcssPlugins(),
      })
      .end()
      .use('less')
      .loader(require.resolve('less-loader'))
      .end();

    config.module
      .rule('sass')
      .test(/\.scss$/)
      .use('style')
      .loader(require.resolve('style-loader'))
      .end()
      .use('css')
      .loader(require.resolve('css-loader'))
      .end()
      .use('postcss')
      .loader(require.resolve('postcss-loader'))
      .options({
        ident: 'postcss',
        plugins: () => getPostcssPlugins(),
      })
      .end()
      .use('sass')
      .loader(require.resolve('sass-loader'))
      .end();
  } else {
    const cssLoaderOptions = {
      modules: {
        auto: /\.module\.\w+$/i,
        localIdentName: '[path][name]__[local]--[hash:base64:5]',
      },
    };

    const postcssOptions = {
      ident: 'postcss',
      plugins: () => [
        require('postcss-preset-env')({
          autoprefixer: {
            flexbox: 'no-2009',
          },
          stage: 3,
        }),
        require('postcss-plugin-rpx2vw')(),
      ],
    };

    config.module
      .rule('css')
      .test(/\.css?$/)
      .use('MiniCssExtractPlugin.loader')
      .loader(MiniCssExtractPlugin.loader)
      .options({
        esModule: false,
      })
      .end()
      .use('css')
      .loader(require.resolve('css-loader'))
      .options(cssLoaderOptions)
      .end()
      .use('postcss')
      .loader(require.resolve('postcss-loader'))
      .options(postcssOptions);

    config.module
      .rule('less')
      .test(/\.less?$/)
      .use('MiniCssExtractPlugin.loader')
      .loader(MiniCssExtractPlugin.loader)
      .options({
        esModule: false,
      })
      .end()
      .use('css-loader')
      .loader(require.resolve('css-loader'))
      .options(cssLoaderOptions)
      .end()
      .use('postcss-loader')
      .loader(require.resolve('postcss-loader'))
      .options(postcssOptions)
      .end()
      .use('less-loader')
      .loader(require.resolve('less-loader'));

    config.module
      .rule('scss')
      .test(/\.scss?$/)
      .use('MiniCssExtractPlugin.loader')
      .loader(MiniCssExtractPlugin.loader)
      .options({
        esModule: false,
      })
      .end()
      .use('css-loader')
      .loader(require.resolve('css-loader'))
      .options(cssLoaderOptions)
      .end()
      .use('postcss-loader')
      .loader(require.resolve('postcss-loader'))
      .options(postcssOptions)
      .end()
      .use('sass-loader')
      .loader(require.resolve('sass-loader'));
  }
};
